{
 "cells": [
  {
   "cell_type": "code",
   "id": "f50ce05f-a75b-4134-b510-f95ecfc06b0c",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "library(GenomicRanges)\n",
    "library(TCGAbiolinks)\n",
    "library(survminer)\n",
    "library(survival)\n",
    "library(SummarizedExperiment)\n",
    "library(tidyverse)\n",
    "library(DESeq2)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "58a5d953-fd1e-4713-b30b-440fd8fd741f",
   "metadata": {},
   "source": [
    "# Set the folder path where the TSV files are located\n",
    "folder_path <- \"/scratch/project/stseq/Onkar/BigData/Melanoma_TCGA_bulk_GE/\"\n",
    "\n",
    "\n",
    "# Get a list of all TSV files in the folder\n",
    "files <- list.files(path = folder_path, pattern = \"\\\\.tsv$\", full.names = TRUE)\n",
    "\n",
    "# Initialize an empty list to store data frames\n",
    "df_list <- list()\n",
    "\n",
    "# Loop through each file and read it into a list, extracting only the tpm_unstranded column\n",
    "for (file in files) {\n",
    "  # Read the TSV file\n",
    "  df <- read.table(file, header = TRUE, sep = \"\\t\")\n",
    "  \n",
    "  # Extract gene_name and tpm_unstranded\n",
    "  gene_name <- df$gene_name\n",
    "  tpm_data <- df$tpm_unstranded\n",
    "  \n",
    "  # Assign the column name as the file name without the \".tsv\" extension\n",
    "  colname <- gsub(\".rna_seq.augmented_star_gene_counts.tsv$\", \"\", basename(file))  # Remove \".tsv\" from file name\n",
    "  \n",
    "  # Add the tpm_unstranded data to the list with the appropriate column name\n",
    "  # Assign gene_name as rownames\n",
    "  df_list[[colname]] <- setNames(tpm_data, gene_name)\n",
    "}\n",
    "\n",
    "# Combine all data frames into one by rows, with gene_name as rownames\n",
    "combined_df <- do.call(cbind, df_list)\n",
    "\n",
    "# View the combined data frame with gene_name as row names\n",
    "head(combined_df)\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "6c097aa7-00f7-4ba9-acf1-e6169075f48d",
   "metadata": {},
   "source": [
    "combined_df<-combined_df[-c(1:4),]\n",
    "head(combined_df)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "40061cbb-fd32-47d8-b485-f246ccbaf6b4",
   "metadata": {},
   "source": [
    "library(dplyr)\n",
    "library(survminer)\n",
    "library(survival)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "6021dc69-4fff-4742-8b97-d7eeb7f62294",
   "metadata": {},
   "source": [
    "coldata<-read.csv(\"TCGA/Mel/Clinical_data_mel.txt\",sep=\"\\t\",header=TRUE)\n",
    "coldata_reordered <- as.data.frame(coldata[match(colnames(combined_df), coldata$V1), ])\n",
    "#names(coldata_reordered)<-c(\"ID\",\"deceased\",\"submitter_id\",\"overall_survival\")\n",
    "coldata_reordered<-coldata_reordered[-(length(rownames(coldata_reordered))),]\n",
    "rownames(coldata_reordered)<-coldata_reordered$ID\n",
    "combined_df<-combined_df[,rownames(coldata_reordered)]\n",
    "combined_df[combined_df == 0] <- 1\n",
    "coldata_reordered$deceased <- ifelse(coldata_reordered$deceased == \"dead\", TRUE, FALSE)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "17f10824-6c1e-4d51-aec9-ab48690563d8",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "88e1aeb7-dcfe-4833-803a-fb97c7a10e40",
   "metadata": {},
   "source": [
    "coldata<-read.csv(\"TCGA/Mel/Clinical_data_mel.txt\",sep=\"\\t\",header=TRUE)\n",
    "names(coldata)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "c5ecbaec-57c2-4238-ba5b-80e87dae280f",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "37aaaa69-64d8-410a-a1b2-7153cc5e3a0b",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "d8f82165-afe3-4c25-8eee-a7225f195b8d",
   "metadata": {},
   "source": [
    "intersect(colnames(combined_df),coldata$submitter_id)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8f9b8bf4-7736-454b-bd5a-6f022b88b5fc",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "head(colnames(combined_df))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bb4dce7f-9ba8-4d91-9511-5414595df60d",
   "metadata": {},
   "source": [
    "setwd(\"Skin_Melanoma_TCGA/\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "0dbd4b3c-0c38-4325-836c-21d6b2026cd6",
   "metadata": {},
   "source": [
    "# Assuming coldata is your data frame with the 'project' column\n",
    "library(dplyr)\n",
    "\n",
    "# Subset coldata for project \"TCGA-SKCM\"\n",
    "coldata_TCGASKCM <- coldata %>%\n",
    "  filter(project == \"TCGA-SKCM\")\n",
    "\n",
    "# View the result\n",
    "head(coldata_TCGASKCM)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e7c5d9fa-a347-464e-9e16-c4c2350e30c0",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# script to download data from TCGA using TCGAbiolinks\n",
    "\n",
    "# get a list of projects\n",
    "gdcprojects <- getGDCprojects()\n",
    "getProjectSummary('TCGA-SKCM')\n",
    "\n",
    "\n",
    "\n",
    "# building a query\n",
    "query_TCGA <- GDCquery(project = 'TCGA-SKCM',\n",
    "         data.category = 'Transcriptome Profiling')\n",
    "output_query_TCGA <- getResults(query_TCGA)\n",
    "\n",
    "\n",
    "# build a query to retrieve gene expression data ------------\n",
    "query_TCGA <- GDCquery(project = 'TCGA-SKCM',\n",
    "                       data.category = 'Transcriptome Profiling',\n",
    "                       experimental.strategy = 'RNA-Seq',\n",
    "                       workflow.type = 'STAR - Counts',\n",
    "                       access = 'open',   data.type = \"Gene Expression Quantification\"\n",
    "                       )\n",
    "\n",
    "getResults(query_TCGA)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8d638029-7cbc-4a90-8fc3-1d5c772fccd7",
   "metadata": {},
   "source": [
    "query_TCGA_meta<-getResults(query_TCGA)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "895685f2-a01d-4397-a8b0-a24e294034b7",
   "metadata": {},
   "source": [
    "names(coldata_TCGASKCM)\n",
    "names(query_TCGA_meta)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8813fd40-1ef2-41a0-8ce2-ce3d16bffac0",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# Merge coldata_TCGASKCM and query_TCGA_meta by submitter_id\n",
    "print(coldata_TCGASKCM[1,])\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "d057e58d-e620-4384-9fc4-f3632658a71d",
   "metadata": {},
   "source": [
    "#grep(\"TCGA-EE-A2GU\", coldata_TCGASKCM$submitter_id, value = TRUE)\n",
    "#grep(\"TCGA-EE-A2GU\", query_TCGA_meta$cases.submitter_id, value = TRUE)\n",
    "\n",
    "grep(\"TCGA-EE-A2GU\", coldata_TCGASKCM, value = FALSE)\n",
    "grep(\"TCGA-FS-A1ZZ\", query_TCGA_meta, value = FALSE)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "30727adf-d0a3-4fb6-8e76-f396944603d7",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "abd96531-56b9-47db-866e-043ff4902864",
   "metadata": {},
   "source": [
    "# Use grep to find the row index of the value \"TCGA-FS-A1ZZ\" in the 'submitter_id' column\n",
    "matching_rows <- grep(\"TCGA-EE-A2GU\", query_TCGA_meta$cases.submitter_id, value = FALSE)\n",
    "\n",
    "# Print the row numbers where it was found\n",
    "matching_rows\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e99d9081-854f-4b02-b89f-46a5c67170a1",
   "metadata": {},
   "source": [
    "diagnosis_id"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "7835c582-2cc1-4747-bf23-3d876e8fd08d",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "query_TCGA_meta$submitter_id"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bb05927c-95d7-483f-a839-d6e2f9b4cbfa",
   "metadata": {},
   "source": [
    "rownames(coldata_TCGASKCM)<-coldata_TCGASKCM$submitter_id\n",
    "\n",
    "\n",
    "# Check for duplicate values in the 'submitter_id' column\n",
    "sum(duplicated(query_TCGA_meta$cases.submitter_id))\n",
    "# View the duplicate submitter_id values\n",
    "query_TCGA_meta$cases.submitter_id[duplicated(query_TCGA_meta$cases.submitter_id)]\n",
    "# Remove duplicates and keep only the first occurrence\n",
    "query_TCGA_meta <- query_TCGA_meta[!duplicated(query_TCGA_meta$cases.submitter_id), ]\n",
    "rownames(query_TCGA_meta)<-query_TCGA_meta$cases.submitter_id\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5e09d5f6-7eb0-4e36-98c7-7c1027beedfa",
   "metadata": {},
   "source": [
    "meta_combined <- merge(coldata_TCGASKCM, query_TCGA_meta, by = \"row.names\", all = FALSE)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "70b335e0-6c67-4e4c-be66-e7f6b04e7e4b",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "rownames(meta_combined)<-meta_combined$Row.names\n",
    "head(meta_combined)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "7f62302b-91ee-49d8-94f4-fa7aeb9bac7f",
   "metadata": {},
   "source": [
    "length(intersect(rownames(meta_combined), rownames(coldata_TCGASKCM)))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "6f605d14-1929-4603-86b2-ee12e6e21596",
   "metadata": {},
   "source": [
    "dim(meta_combined)\n",
    "dim(coldata_TCGASKCM)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "2886248a-fe35-41d7-8753-9547c79af453",
   "metadata": {},
   "source": [
    "# Ensure rownames of meta_combined are set (assuming 'file_id' is the matching column)\n",
    "rownames(meta_combined) <- meta_combined$file_id\n",
    "\n",
    "# Get the mapping of file_id to row names\n",
    "new_colnames <- rownames(meta_combined)[match(colnames(combined_df), meta_combined$file_id)]\n",
    "\n",
    "# Replace the column names of combined_df with the matched row names\n",
    "colnames(combined_df) <- new_colnames\n",
    "\n",
    "# Check the renamed columns in combined_df\n",
    "head(combined_df)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9226cb73-247f-4d57-b944-4d90418b9a20",
   "metadata": {},
   "source": [
    "head(colnames(combined_df))\n",
    "head(meta_combined$file_id )"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "cf81259c-4b36-4ab8-a380-b252a520d064",
   "metadata": {},
   "source": [
    "length(intersect(colnames(combined_df),meta_combined$file_id))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "fb127080-390a-4a7d-8ecb-0ce7e8892bd0",
   "metadata": {},
   "source": [
    "# Clean up both colnames and file_id by removing extra spaces\n",
    "colnames(combined_df) <- trimws(colnames(combined_df))\n",
    "meta_combined$file_id <- trimws(meta_combined$file_id)\n",
    "\n",
    "# Check if there are still any leading/trailing spaces\n",
    "intersect(colnames(combined_df), meta_combined$file_id)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5303b152-0f35-4ff4-bd9c-74e82177ab01",
   "metadata": {},
   "source": [
    "head(meta_combined)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "57156c26-29c6-4a8e-8683-0ed710379a42",
   "metadata": {},
   "source": [
    "meta_combined$analysis_submitter_id<-gsub(\"_star__counts\",\"\",meta_combined$analysis_submitter_id)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "2490e683-4a8a-43e7-94fe-fb47bc4a4e04",
   "metadata": {},
   "source": [
    "combined_df_OG<-combined_df\n",
    "# Ensure rownames of meta_combined are set (assuming 'file_id' is the matching column)\n",
    "rownames(meta_combined) <- meta_combined$Row.names\n",
    "\n",
    "# Get the mapping of file_id to row names\n",
    "new_colnames <- rownames(meta_combined)[match(colnames(combined_df), meta_combined$analysis_submitter_id)]\n",
    "\n",
    "# Replace the column names of combined_df with the matched row names\n",
    "colnames(combined_df) <- new_colnames\n",
    "\n",
    "# Check the renamed columns in combined_df\n",
    "head(combined_df)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "45efeef7-9e87-42de-9514-0349ec58af77",
   "metadata": {},
   "source": [
    "rownames(coldata)<-coldata$submitter_id\n",
    "# Convert to characters if they are factors\n",
    "colnames(combined_df) <- as.character(colnames(combined_df))\n",
    "rownames(coldata) <- as.character(rownames(coldata))\n",
    "\n",
    "# Get the intersect of column names of combined_df and row names of coldata\n",
    "matching_names <- intersect(colnames(combined_df), rownames(coldata))\n",
    "\n",
    "# Subset combined_df based on the matching names\n",
    "combined_df_subset <- combined_df[, matching_names]\n",
    "\n",
    "# View the result to check the subsetted data\n",
    "head(combined_df_subset)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9b365647-c7b4-4a70-95cb-7a803b891bc2",
   "metadata": {},
   "source": [
    "dim(combined_df_subset)\n",
    "dim(coldata)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b844a792-6080-4bfe-a3b0-630bde60bfac",
   "metadata": {},
   "source": [
    "# Subset combined_df based on the matching names\n",
    "coldata_subset <- coldata[matching_names, ]\n",
    "\n",
    "# View the result to check the subsetted data\n",
    "dim(coldata_subset)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b6383014-33c2-4401-86cc-17c1c7d07bd5",
   "metadata": {},
   "source": [
    "# NOw use counts from combined_df_subset and meta from coldata_subset\n",
    "write.table(combined_df_subset,\"/scratch/project_mnt/S0010/Prakrithi/Skin_Melanoma_TCGA/SKCM_tpm_mat.txt\",sep=\"\\t\", quote = FALSE)\n",
    "write.table(coldata_subset,\"/scratch/project_mnt/S0010/Prakrithi/Skin_Melanoma_TCGA/SKCM_meta.txt\",sep=\"\\t\", quote = FALSE)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "aa12d958-16e9-4c63-92d3-ecff1c4d76ab",
   "metadata": {},
   "source": [
    "head(colnames(combined_df_subset))\n",
    "head(rownames(coldata_subset))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ab2b04ec-f66b-4a5d-b97c-f03b4dba78ad",
   "metadata": {},
   "source": [
    "# read while rerunning\n",
    "combined_df_subset<-read.csv(\"/scratch/project_mnt/S0010/Prakrithi/Skin_Melanoma_TCGA/SKCM_tpm_mat.txt\", sep=\"\\t\", header=TRUE, check.names=FALSE)\n",
    "coldata_subset<-read.csv(\"/scratch/project_mnt/S0010/Prakrithi/Skin_Melanoma_TCGA/SKCM_meta.txt\",sep=\"\\t\")\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "61ce40f3-83aa-452e-ad2f-f5562f5deb7c",
   "metadata": {},
   "source": [
    "\n",
    "# Remove duplicate gene entries, keeping only the first occurrence\n",
    "combined_df_subset <- combined_df_subset[!duplicated(combined_df_subset$GENE), ]\n",
    "\n",
    "# Set row names as the GENE column\n",
    "rownames(combined_df_subset) <- combined_df_subset$GENE\n",
    "\n",
    "# Optionally, remove the GENE column if it's now redundant\n",
    "combined_df_subset$GENE <- NULL\n",
    "head(combined_df_subset)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "2dfc9c23-133d-45ff-ac09-655376c8d6de",
   "metadata": {},
   "source": [
    "\n",
    "rounded_matrix <- round(combined_df_subset, digits = 0)\n",
    "head(rounded_matrix)\n",
    "\n",
    "# Check if there are any NA values in the entire data frame\n",
    "any(is.na(rounded_matrix))\n",
    "# Count the number of NA values in the entire data frame\n",
    "sum(is.na(rounded_matrix))\n",
    "# Find which rows in a specific column are NA (e.g., 'column_name')\n",
    "which(is.na(rounded_matrix))\n",
    "# Replace all NA values with 0 in the entire data frame\n",
    "rounded_matrix[is.na(rounded_matrix)] <- 0\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "3344933d-41ac-465a-ab10-33dc31c3bb3d",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "969c5f2d-18c6-4db4-a425-403b6e59695c",
   "metadata": {},
   "source": [
    "dds <- DESeqDataSetFromMatrix(countData = as.matrix(rounded_matrix),\n",
    "                              colData = coldata_subset,\n",
    "                              design = ~ 1)\n",
    "\n",
    "\n",
    "vsd <- vst(dds, blind = TRUE,fitType='parametric') #nsub = nrow(dds\n",
    "matrix_vst <- assay(vsd)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bc6302fe-263a-492c-ab93-d8ba98979cfc",
   "metadata": {},
   "source": [
    "# Define the list of genes you want to stratify by\n",
    "genes_to_stratify <- c(\"THSD8\",\"STC2\",\"CHRFAM7A\",\"AK5\",\"SLC12A3\",\"MED12L\",\"PLAC8\",\"NME8\",\"CLEC4C\",\"PACSIN1\",\"LILRA4\",\"C5AR2\",\"CXCL2\",\"TREML2\",\"IRF8\",\"AGT\",\"WDFY4\",\"NCR1\",\"SLC9A3\",\"LYPD1\",\"KLRC1\",\"KLRC3\",\"ITGAD\",\"ADAMDEC1\",\"FCRL3\",\"IL18RAP\",\"HORMAD1\",\"ANGPT1\",\"BMP8B\",\"DLX1\",\"LRP2\",\"ANO4\",\"UPK1A\",\"COL22A1\",\"ITGA10\",\"STRA6\",\"CD80\",\"BMX\",\"CABP4\",\"NKG7\",\"SH2D1A\",\"DTHD1\",\"SELL\",\"IL2RB\",\"UBD\",\"MPZ\",\"CXCL9\",\"CTLA4\",\"P2RY10\",\"IL3RA\")  # Replace with your list of genes\n",
    "\n",
    "# Create a new dataframe to store the stratified data\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = 'gene_id') %>%\n",
    "  gather(key = 'case_id', value = 'counts', -gene_id)\n",
    "\n",
    "# Filter to include only the genes in the list\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% genes_to_stratify)\n",
    "\n",
    "# Calculate the combined expression for each sample\n",
    "combined_expression <- filtered %>%\n",
    "  spread(key = 'gene_id', value = 'counts') %>%\n",
    "  select(-case_id) %>%\n",
    "  rowSums()\n",
    "\n",
    "# Add the combined expression to the dataframe\n",
    "combined$combined_expression <- combined_expression[match(combined$case_id, filtered$case_id)]\n",
    "\n",
    "# Calculate the median of the combined expression across all samples\n",
    "median_combined_expression <- median(combined$combined_expression, na.rm = TRUE)\n",
    "\n",
    "# Initialize a new strata column based on the combined median expression\n",
    "combined$strata <- ifelse(combined$combined_expression >= median_combined_expression, \"HIGH\", \"LOW\")\n",
    "\n",
    "# Merge the clinical data to get survival information\n",
    "combined <- merge(combined, coldata_subset, by.x = 'case_id', by.y = 'submitter_id')\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit the survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Plot the Kaplan-Meier curve\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"))\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "dc457928-52f1-4ab4-a75f-e3565f0549bc",
   "metadata": {},
   "source": [
    "grep(\"THSD8\",rownames(rounded_matrix))\n",
    "rounded_matrix[56334,]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "49005426-f308-42f5-a5cc-5e0ccf06f15b",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8a69a691-f6ae-49f6-945a-abffd35340fd",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "cd673082-5360-447a-adb0-38315383b18c",
   "metadata": {},
   "source": [
    "# NEW ANALYSIS (CORRECT)"
   ]
  },
  {
   "cell_type": "code",
   "id": "d3d3c289-1c45-4415-9c68-c700a0a4c381",
   "metadata": {},
   "source": [
    "# read while rerunning\n",
    "combined_df_subset<-read.csv(\"/scratch/project_mnt/S0010/Prakrithi/Skin_Melanoma_TCGA/SKCM_tpm_mat.txt\", sep=\"\\t\", header=TRUE, check.names=FALSE)\n",
    "coldata_subset<-read.csv(\"/scratch/project_mnt/S0010/Prakrithi/Skin_Melanoma_TCGA/SKCM_meta.txt\",sep=\"\\t\")\n",
    "\n",
    "\n",
    "# Remove duplicate gene entries, keeping only the first occurrence\n",
    "combined_df_subset <- combined_df_subset[!duplicated(combined_df_subset$GENE), ]\n",
    "\n",
    "# Set row names as the GENE column\n",
    "rownames(combined_df_subset) <- combined_df_subset$GENE\n",
    "\n",
    "# Optionally, remove the GENE column if it's now redundant\n",
    "combined_df_subset$GENE <- NULL\n",
    "head(combined_df_subset)\n",
    "\n",
    "rounded_matrix <- round(combined_df_subset, digits = 0)\n",
    "head(rounded_matrix)\n",
    "\n",
    "# Check if there are any NA values in the entire data frame\n",
    "any(is.na(rounded_matrix))\n",
    "# Count the number of NA values in the entire data frame\n",
    "sum(is.na(rounded_matrix))\n",
    "# Find which rows in a specific column are NA (e.g., 'column_name')\n",
    "which(is.na(rounded_matrix))\n",
    "# Replace all NA values with 0 in the entire data frame\n",
    "rounded_matrix[is.na(rounded_matrix)] <- 0\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4c8fee99-6049-4ed6-8c56-03d001bc6458",
   "metadata": {},
   "source": [
    "dds <- DESeqDataSetFromMatrix(countData = as.matrix(rounded_matrix),\n",
    "                              colData = coldata_subset,\n",
    "                              design = ~ 1)\n",
    "\n",
    "\n",
    "vsd <- vst(dds, blind = TRUE,fitType='parametric') #nsub = nrow(dds\n",
    "matrix_vst <- assay(vsd)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "176b08d9-1ff2-4e65-a9c2-7d855bffb64a",
   "metadata": {},
   "source": [
    "# DE genes\n",
    "genes_to_stratify <- c(\"THSD8\", \"STC2\", \"CHRFAM7A\", \"AK5\", \"SLC12A3\", \"MED12L\", \"PLAC8\", \"NME8\", \n",
    "                        \"CLEC4C\", \"PACSIN1\", \"LILRA4\", \"C5AR2\", \"CXCL2\", \"TREML2\", \"IRF8\", \"AGT\", \n",
    "                        \"WDFY4\", \"NCR1\", \"SLC9A3\", \"LYPD1\", \"KLRC1\", \"KLRC3\", \"ITGAD\", \"ADAMDEC1\", \n",
    "                        \"FCRL3\", \"IL18RAP\", \"HORMAD1\", \"ANGPT1\", \"BMP8B\", \"DLX1\", \"LRP2\", \"ANO4\", \n",
    "                        \"UPK1A\", \"COL22A1\", \"ITGA10\", \"STRA6\", \"CD80\", \"BMX\", \"CABP4\", \"NKG7\", \n",
    "                        \"SH2D1A\", \"DTHD1\", \"SELL\", \"IL2RB\", \"UBD\", \"MPZ\", \"CXCL9\", \"CTLA4\", \"P2RY10\", \n",
    "                        \"IL3RA\")\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% genes_to_stratify)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Plot Kaplan-Meier curve\n",
    "pdf(\"Skin_Melanoma_TCGA/TCGA_mel_DE_genes.pdf\", width = 8, height = 9)\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "dev.off()\n",
    "\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ed71d9e3-2abe-4c06-b533-d855d4553e9d",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# Ensure genes_to_stratify exist in matrix_vst\n",
    "filtered <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id) %>%\n",
    "  filter(gene_id %in% genes_to_stratify) %>%\n",
    "  spread(key = \"gene_id\", value = \"counts\")\n",
    "\n",
    "# Merge with combined dataframe\n",
    "combined <- merge(combined, filtered, by.x = \"case_id\", by.y = \"case_id\", all.x = TRUE)\n",
    "\n",
    "combined <- combined %>%\n",
    "  mutate(status = ifelse(strata == \"HIGH\", 1, 0))\n",
    "# Fit Cox proportional hazards model for Gene1\n",
    "formula_string <- paste(\"Surv(overall_survival, status) ~\", paste(genes_to_stratify, collapse = \" + \"))\n",
    "cox_fit <- coxph(as.formula(formula_string), data = combined)\n",
    "\n",
    "# Print results\n",
    "summary(cox_fit)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "328f5287-bd92-402a-bb22-9a54a00b9a68",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "0459df51-5038-4bc4-a586-9c58bd35f846",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "e27f9f22-7a72-4490-aadd-6b003f1bce65",
   "metadata": {},
   "source": [
    "# LR genes"
   ]
  },
  {
   "cell_type": "code",
   "id": "96ca82bd-44ca-45c4-87b5-7e658826f7e7",
   "metadata": {},
   "source": [
    "LRgenes<-c(\"COL1A1\",\"DDR2\",\"COL1A1\",\"ITGB1\",\"COL1A1\",\"CD44\",\"COL1A1\",\"CD36\",\"COL1A1\",\"ITGA5\",\"COL1A1\",\"ITGA2\",\"COL1A1\",\"DDR1\",\"COL3A1\",\"DDR2\",\"COL3A1\",\"DDR1\",\"COL1A2\",\"ITGB1\",\"COL1A2\",\"ITGA2\",\"COL1A2\",\"CD44\",\"CCL19\",\"CXCR3\",\"CCL5\",\"ACKR1\",\"AZGP1\",\"ITGAV\",\"CCL5\",\"CCR1\",\"FGF1\",\"FGFR1\",\"IL34\",\"CSF1R\",\"ICAM3\",\"ITGB2\",\"CD34\",\"SELL\",\"FGF1\",\"CD44\",\"FGF2\",\"CD44\",\"COL8A1\",\"ITGA1\",\"LGALS3\",\"LAG3\",\"NRG1\",\"ERBB3\",\"TNFSF14\",\"TNFRSF14\")\n",
    "# Create a new dataframe to store the stratified data\n",
    "\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% LRgenes)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Plot Kaplan-Meier curve\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "\n",
    "pdf(\"Skin_Melanoma_TCGA/TCGA_mel_LR_genes.pdf\", width = 8, height = 9)\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "dev.off()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "f331265a-37d8-4329-b6e4-593eb2c4229b",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# Ensure genes_to_stratify exist in matrix_vst\n",
    "filtered <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id) %>%\n",
    "  filter(gene_id %in% LRgenes) %>%\n",
    "  spread(key = \"gene_id\", value = \"counts\")\n",
    "\n",
    "# Merge with combined dataframe\n",
    "combined <- merge(combined, filtered, by.x = \"case_id\", by.y = \"case_id\", all.x = TRUE)\n",
    "\n",
    "combined <- combined %>%\n",
    "  mutate(status = ifelse(strata == \"HIGH\", 1, 0))\n",
    "# Fit Cox proportional hazards model for Gene1\n",
    "formula_string <- paste(\"Surv(overall_survival, status) ~\", paste(LRgenes, collapse = \" + \"))\n",
    "cox_fit <- coxph(as.formula(formula_string), data = combined)\n",
    "\n",
    "# Print results\n",
    "summary(cox_fit)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ebd42737-ffbf-46f7-90e9-d500e4191db0",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "1d20c900-ac2e-489f-a7ca-dfb5e260e23a",
   "metadata": {},
   "source": [
    "gsmap_genes_mel4<-c(\"GTF2F2\",\"WDFY1\",\"PBRM1\",\"BRD4\",\"HGS\",\"WDR82\",\"DPP9\",\"CBX3\",\"RABGGTB\",\"POLR2G\",\"CLPTM1L\",\"PSMB2\",\"MED15\",\"UBE2R2\",\"APH1A\",\"TBCD\",\"PSMA1\",\"CDK4\",\"CSNK1D\",\"CCT5\",\"DDX24\",\"VPS29\",\"OAS1\",\"M6PR\",\"EFTUD2\",\"CNDP2\",\"FKBP3\",\"PSMD4\",\"MLLT10\",\"DDX41\",\"SUPT16H\",\"SPPL2A\",\"ATOX1\",\"SEPTIN9\",\"MXD4\",\"EIF3B\",\"SAP30BP\",\"SLC11A2\",\"STAT2\",\"ATP6V0D1\",\"VMP1\",\"UBQLN1\",\"KPNB1\",\"BANF1\",\"ACIN1\",\"AKT2\",\"SLTM\",\"NARF\",\"CIAO1\",\"NPEPL1\")\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% gsmap_genes_mel4)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Plot Kaplan-Meier curve\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b7e2cd1a-b0d4-427f-932f-055152c42948",
   "metadata": {},
   "source": [
    "gsmap_top50all_mel4<-c(\"GTF2F2\",\"PSME2\",\"WDFY1\",\"PBRM1\",\"BRD4\",\"HGS\",\"WDR82\",\"DPP9\",\"CBX3\",\"PML\",\"RABGGTB\",\"POLR2G\",\"CLPTM1L\",\"PSMB2\",\"MED15\",\"UBE2R2\",\"XAF1\",\"APH1A\",\"TBCD\",\"PSMA1\",\"TAP1\",\"CDK4\",\"CSNK1D\",\"CCT5\",\"DDX24\",\"NFKB2\",\"VPS29\",\"OAS1\",\"M6PR\",\"C6orf62\",\"MOV10\",\"CDC42SE1\",\"EFTUD2\",\"IFI16\",\"LARP4B\",\"CNDP2\",\"FKBP3\",\"PSMD4\",\"CMTM6\",\"MLLT10\",\"TAP2\",\"DDX41\",\"CHKB\",\"SUPT16H\",\"SPPL2A\",\"UBP1\",\"SCO2\",\"ATOX1\",\"SEPTIN9\",\"MXD4\")\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% gsmap_top50all_mel4)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Plot Kaplan-Meier curve\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4a0cc586-7955-422e-bc5a-a4cbf8bb4e25",
   "metadata": {},
   "source": [
    "gsmap_common_mel<-c(\"WDFY1\",\"PBRM1\",\"TBCD\",\"CDK4\",\"M6PR\",\"EFTUD2\",\"PSMD4\",\"MXD4\",\"STAT2\",\"PFKL\",\"MAP3K11\",\"MXI1\",\"EIF4E2\",\"SLC39A1\",\"PSMA7\",\"ATP6V1B2\",\"AC015802.6\",\"HSPE1\",\"PLSCR1\",\"KCTD20\",\"UROD\",\"TRA2B\",\"RBM6\",\"ST6GALNAC2\",\"DR1\",\"PPP4R2\",\"ANKLE2\",\"SMARCD2\",\"OSBPL9\",\"MLEC\",\"SRPRA\",\"HDGF\",\"NELFE\",\"UBXN6\",\"GRSF1\",\"COMMD4\",\"TXNL4A\",\"CAPZA2\",\"ILVBL\",\"UBAP2L\",\"CFAP97\",\"PA2G4\",\"USP22\",\"GPAT4\",\"FBXW5\",\"PTTG1IP\",\"GADD45GIP1\",\"COPG1\",\"ZNF106\",\"WBP2\")\n",
    "\n",
    "#gsmap_top50all_mel4<-c(\"GTF2F2\",\"PSME2\",\"WDFY1\",\"PBRM1\",\"BRD4\",\"HGS\",\"WDR82\",\"DPP9\",\"CBX3\",\"PML\",\"RABGGTB\",\"POLR2G\",\"CLPTM1L\",\"PSMB2\",\"MED15\",\"UBE2R2\",\"XAF1\",\"APH1A\",\"TBCD\",\"PSMA1\",\"TAP1\",\"CDK4\",\"CSNK1D\",\"CCT5\",\"DDX24\",\"NFKB2\",\"VPS29\",\"OAS1\",\"M6PR\",\"C6orf62\",\"MOV10\",\"CDC42SE1\",\"EFTUD2\",\"IFI16\",\"LARP4B\",\"CNDP2\",\"FKBP3\",\"PSMD4\",\"CMTM6\",\"MLLT10\",\"TAP2\",\"DDX41\",\"CHKB\",\"SUPT16H\",\"SPPL2A\",\"UBP1\",\"SCO2\",\"ATOX1\",\"SEPTIN9\",\"MXD4\")\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% gsmap_common_mel)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "\n",
    "pdf(\"Skin_Melanoma_TCGA/TCGA_mel_gsmap_comm_genes.pdf\", width = 8, height = 9)\n",
    "# Plot Kaplan-Meier curve\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "dev.off()\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "a81a30c9-6df1-4950-b0e4-c29b8ccdc789",
   "metadata": {},
   "source": [
    "## gsmap all CT top common\n",
    "\n",
    "gsmap_all_top_comm<-c(\"WDFY1\",\"PBRM1\",\"PML\",\"TBCD\",\"CDK4\",\"M6PR\",\"EFTUD2\",\"IFI16\",\"PSMD4\",\"MXD4\",\"STAT2\",\"PFKL\",\"MAP3K11\",\"C21orf91\",\"MXI1\",\"EIF4E2\",\"NDUFV1\",\"SLC39A1\",\"SNRNP200\",\"PSMA7\",\"ATP6V1B2\",\"AC015802.6\",\"HSPE1\",\"CLASRP\",\"PLSCR1\",\"KCTD20\",\"UROD\",\"RHOT2\",\"TRA2B\",\"RBM6\",\"ST6GALNAC2\",\"DR1\",\"RTRAF\",\"PPP4R2\",\"ANKLE2\",\"SMARCD2\",\"OSBPL9\",\"MLEC\",\"SRPRA\",\"HDGF\",\"ETS1\",\"NELFE\",\"UBXN6\",\"GRSF1\",\"COMMD4\",\"TXNL4A\",\"CAPZA2\",\"SRRT\",\"ILVBL\",\"UBAP2L\")\n",
    "\n",
    "\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% gsmap_all_top_comm)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "\n",
    "\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "84156008-e422-4231-9d9f-00392eb5cac7",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "d0d0dc3d-15fc-4286-8a14-abbf4b8bc037",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b2789bae-03aa-4a19-8f6d-34c0b3b09c19",
   "metadata": {},
   "source": [
    "genes_to_stratify <- c(\"THSD8\", \"STC2\", \"CHRFAM7A\", \"AK5\", \"SLC12A3\", \"MED12L\", \"PLAC8\", \"NME8\", \n",
    "                        \"CLEC4C\", \"PACSIN1\", \"LILRA4\", \"C5AR2\", \"CXCL2\", \"TREML2\", \"IRF8\", \"AGT\", \n",
    "                        \"WDFY4\", \"NCR1\", \"SLC9A3\", \"LYPD1\", \"KLRC1\", \"KLRC3\", \"ITGAD\", \"ADAMDEC1\", \n",
    "                        \"FCRL3\", \"IL18RAP\", \"HORMAD1\", \"ANGPT1\", \"BMP8B\", \"DLX1\", \"LRP2\", \"ANO4\", \n",
    "                        \"UPK1A\", \"COL22A1\", \"ITGA10\", \"STRA6\", \"CD80\", \"BMX\", \"CABP4\", \"NKG7\", \n",
    "                        \"SH2D1A\", \"DTHD1\", \"SELL\", \"IL2RB\", \"UBD\", \"MPZ\", \"CXCL9\", \"CTLA4\", \"P2RY10\", \n",
    "                        \"IL3RA\")\n",
    "length(genes_to_stratify)\n",
    "LRgenes<-c(\"COL1A1\",\"DDR2\",\"COL1A1\",\"ITGB1\",\"COL1A1\",\"CD44\",\"COL1A1\",\"CD36\",\"COL1A1\",\"ITGA5\",\"COL1A1\",\"ITGA2\",\"COL1A1\",\"DDR1\",\"COL3A1\",\"DDR2\",\"COL3A1\",\"DDR1\",\"COL1A2\",\"ITGB1\",\"COL1A2\",\"ITGA2\",\"COL1A2\",\"CD44\",\"CCL19\",\"CXCR3\",\"CCL5\",\"ACKR1\",\"AZGP1\",\"ITGAV\",\"CCL5\",\"CCR1\",\"FGF1\",\"FGFR1\",\"IL34\",\"CSF1R\",\"ICAM3\",\"ITGB2\",\"CD34\",\"SELL\",\"FGF1\",\"CD44\",\"FGF2\",\"CD44\",\"COL8A1\",\"ITGA1\",\"LGALS3\",\"LAG3\",\"NRG1\",\"ERBB3\",\"TNFSF14\",\"TNFRSF14\")\n",
    "length(LRgenes)\n",
    "\n",
    "gsmap_genes_mel4<-c(\"GTF2F2\",\"WDFY1\",\"PBRM1\",\"BRD4\",\"HGS\",\"WDR82\",\"DPP9\",\"CBX3\",\"RABGGTB\",\"POLR2G\",\"CLPTM1L\",\"PSMB2\",\"MED15\",\"UBE2R2\",\"APH1A\",\"TBCD\",\"PSMA1\",\"CDK4\",\"CSNK1D\",\"CCT5\",\"DDX24\",\"VPS29\",\"OAS1\",\"M6PR\",\"EFTUD2\",\"CNDP2\",\"FKBP3\",\"PSMD4\",\"MLLT10\",\"DDX41\",\"SUPT16H\",\"SPPL2A\",\"ATOX1\",\"SEPTIN9\",\"MXD4\",\"EIF3B\",\"SAP30BP\",\"SLC11A2\",\"STAT2\",\"ATP6V0D1\",\"VMP1\",\"UBQLN1\",\"KPNB1\",\"BANF1\",\"ACIN1\",\"AKT2\",\"SLTM\",\"NARF\",\"CIAO1\",\"NPEPL1\")\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "7258fb6d-0eed-4790-acaf-449e15fa1fe6",
   "metadata": {},
   "source": [
    "# Get all genes that are NOT in the two lists\n",
    "excluded_genes <- union(genes_to_stratify, LRgenes)\n",
    "excluded_genes2<-union(gsmap_genes_mel4,gsmap_common_mel)\n",
    "excluded_genes<-union(excluded_genes,excluded_genes2)\n",
    "#excluded_genes<-union(excluded_genes,gsmap_genes_mel4)\n",
    "\n",
    "available_genes <- setdiff(rownames(matrix_vst), excluded_genes)\n",
    "\n",
    "# Randomly select 50 genes from the remaining ones\n",
    "#set.seed(123)  # For reproducibility\n",
    "set.seed(18090909)\n",
    "random_genes <- sample(available_genes, 50)\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the randomly selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% random_genes)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "pdf(\"Skin_Melanoma_TCGA/TCGA_mel_random_exluding_DE_LR_gsmap.pdf\", width = 8, height = 9)\n",
    "# Plot Kaplan-Meier curve\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "dev.off()\n",
    "\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "d3e74981-07c3-4d02-8b67-2fa4ffff624e",
   "metadata": {},
   "source": [
    "available_genes <- setdiff(rownames(matrix_vst), gsmap_common_mel)#gsmap_genes_mel4,gsmap_common_mel\n",
    "\n",
    "# Randomly select 50 genes from the remaining ones\n",
    "#set.seed(123)  # For reproducibility\n",
    "set.seed(0)\n",
    "random_genes <- sample(available_genes, 1)\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the randomly selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% random_genes)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8297c488-a22e-4556-891f-9de53dfa7136",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# Ensure genes_to_stratify exist in matrix_vst\n",
    "filtered <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id) %>%\n",
    "  filter(gene_id %in% LRgenes) %>%\n",
    "  spread(key = \"gene_id\", value = \"counts\")\n",
    "\n",
    "# Merge with combined dataframe\n",
    "combined <- merge(combined, filtered, by.x = \"case_id\", by.y = \"case_id\", all.x = TRUE)\n",
    "\n",
    "combined <- combined %>%\n",
    "  mutate(status = ifelse(strata == \"HIGH\", 1, 0))\n",
    "# Fit Cox proportional hazards model for Gene1\n",
    "formula_string <- paste(\"Surv(overall_survival, status) ~\", paste(LRgenes, collapse = \" + \"))\n",
    "cox_fit <- coxph(as.formula(formula_string), data = combined)\n",
    "\n",
    "# Print results\n",
    "summary(cox_fit)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "61631f41-2913-48a5-b6c9-2c1ba8d04ab0",
   "metadata": {},
   "source": [
    "# Get all genes that are NOT in the two lists\n",
    "#excluded_genes <- union(genes_to_stratify, LRgenes)\n",
    "#available_genes <- setdiff(rownames(matrix_vst), excluded_genes)\n",
    "\n",
    "# Randomly select 50 genes from the remaining ones\n",
    "set.seed(0000000)  # For reproducibility\n",
    "random_genes <- sample(available_genes, 50)\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the randomly selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% random_genes)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Plot Kaplan-Meier curve\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "58f02b85-f894-4b4c-be99-93c388dbea95",
   "metadata": {},
   "source": [
    "# Get all genes that are NOT in the two lists\n",
    "excluded_genes <- union(genes_to_stratify, LRgenes)\n",
    "available_genes <- setdiff(rownames(matrix_vst), excluded_genes)\n",
    "\n",
    "# Randomly select 50 genes from the remaining ones\n",
    "set.seed(100)  # For reproducibility\n",
    "random_genes <- sample(available_genes, 50)\n",
    "\n",
    "# Convert the matrix to a tidy dataframe\n",
    "combined <- matrix_vst %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(var = \"gene_id\") %>%\n",
    "  gather(key = \"case_id\", value = \"counts\", -gene_id)\n",
    "\n",
    "# Filter for the randomly selected genes\n",
    "filtered <- combined %>%\n",
    "  filter(gene_id %in% random_genes)\n",
    "\n",
    "# Compute **mean** expression per sample\n",
    "mean_expression_per_sample <- filtered %>%\n",
    "  group_by(case_id) %>%\n",
    "  summarise(mean_expression = mean(counts, na.rm = TRUE))\n",
    "\n",
    "# Compute the median expression across all samples\n",
    "median_expression <- median(mean_expression_per_sample$mean_expression, na.rm = TRUE)\n",
    "\n",
    "# Assign strata based on median expression\n",
    "mean_expression_per_sample <- mean_expression_per_sample %>%\n",
    "  mutate(strata = ifelse(mean_expression >= median_expression, \"HIGH\", \"LOW\"))\n",
    "\n",
    "# Merge clinical data\n",
    "combined <- merge(mean_expression_per_sample, coldata_subset, by.x = \"case_id\", by.y = \"submitter_id\")\n",
    "\n",
    "# Perform survival analysis\n",
    "survdiff_res <- survdiff(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "# Fit survival model\n",
    "fit <- survfit(Surv(overall_survival) ~ strata, data = combined)\n",
    "\n",
    "\n",
    "pdf(\"Skin_Melanoma_TCGA/TCGA_mel_random_exluding_DE_LR.pdf\", width = 8, height = 9)\n",
    "# Plot Kaplan-Meier curve\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n",
    "dev.off()\n",
    "\n",
    "ggsurvplot(fit,\n",
    "           data = combined,\n",
    "           pval = TRUE,\n",
    "           risk.table = TRUE,\n",
    "           legend.title = \"Strata\",\n",
    "           legend.labs = c(\"LOW\", \"HIGH\"),\n",
    "           palette = c(\"blue\", \"red\"))  # LOW = blue, HIGH = red\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "807025e7-a6df-4fb7-a2e2-c0e03edcc78d",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R_TCGA",
   "language": "R",
   "name": "r_tcga"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
