{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e9d1bf54",
   "metadata": {},
   "source": [
    "## edgeR pipeline followed for filtering genes and normalising gene-expression\n",
    "1. Unnormalised and Normalised gene expression boxplots \n",
    "2. MDS Plots\n",
    "3. Gene/Protein expression heatmaps \n",
    "\n",
    "## Spatial Deconvolution\n",
    "1. Reference matrix heatmap\n",
    "2. Cell type deconvolution barplot\n",
    "3. Cell type proportion difference in PanCK vs Stroma boxplot "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "826530c1",
   "metadata": {},
   "source": [
    "# Loading Libraries"
   ]
  },
  {
   "cell_type": "code",
   "id": "5a6feb9c",
   "metadata": {},
   "source": [
    "#################################################################################################################################################\n",
    "# Load required libraries\n",
    "#################################################################################################################################################\n",
    "\n",
    "library(readxl)\n",
    "library(patchwork)\n",
    "library(reshape2)\n",
    "library(SpatialDecon)\n",
    "library(clusterSim)\n",
    "library(edgeR)\n",
    "library(limma)\n",
    "library(Glimma)\n",
    "library(org.Mm.eg.db)\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "library(stringr)\n",
    "library(reticulate)\n",
    "library(janitor)\n",
    "library(tibble)\n",
    "library(UpSetR)\n",
    "library(factoextra)\n",
    "library(gplots)\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "library(RColorBrewer)\n",
    "library(pheatmap)\n",
    "options(warn=-1)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "704855f4",
   "metadata": {},
   "source": [
    "# Loading Data"
   ]
  },
  {
   "cell_type": "code",
   "id": "f3479118",
   "metadata": {},
   "source": [
    "#################################################################################################################################################\n",
    "# Load Data and Metadata\n",
    "#################################################################################################################################################\n",
    "\n",
    "#Path = \"/QRISdata/Q2051/NanoString_SCC/\"\n",
    "Path = \"D:/Onkar_D/UQ/Project_NanoString Single Cell/NanoString_SCC/csvFiles/\"\n",
    "\n",
    "data(\"safeTME\")\n",
    "data(\"safeTME.matches\")\n",
    "data(\"cellcols\")\n",
    "\n",
    "all_cancer_PatientOverlapResults <- read.delim(paste0(Path,\"all_cancer_PatientOverlapResults.txt\"))\n",
    "all_normal_PatientOverlapResults <- read.delim(paste0(Path,\"all_normal_PatientOverlapResults.txt\"))\n",
    "CTA_Bioprobe_QC <- read_excel(paste0(Path,\"All Data CTA.xlsx\"), sheet = \"Q3\")\n",
    "Protein_QC <- read_excel(paste0(Path,\"All Data Human IO Protein.xlsx\"), sheet = \"HouseKeeping Normalized\")\n",
    "Segment_Properties <- read.csv(paste0(Path,\"Segment_Properties.csv\"))\n",
    "Segment_Properties <- Segment_Properties[seq_len(nrow(Segment_Properties)) + c(1,-1),]\n",
    "\n",
    "\n",
    "#################################################################################################################################################\n",
    "# RNA Data Wrangling\n",
    "#################################################################################################################################################\n",
    "\n",
    "CTA_Bioprobe_QC$ScanLabel <- substring(CTA_Bioprobe_QC$ScanLabel, 5)\n",
    "CTA_Bioprobe_QC$SegmentLabel <- paste(CTA_Bioprobe_QC$SegmentLabel, CTA_Bioprobe_QC$ROILabel, CTA_Bioprobe_QC$ScanLabel)\n",
    "CTA_Bioprobe_QC_SC <- CTA_Bioprobe_QC[,-c(1:2)] \n",
    "CTA_Bioprobe_QC_SC$SegmentLabel <- gsub(\" \", \"_\", CTA_Bioprobe_QC_SC$SegmentLabel)\n",
    "CTA_Bioprobe_QC_SC <- as.data.frame(CTA_Bioprobe_QC_SC)\n",
    "CTA_Bioprobe_QC_SC <- data.frame(CTA_Bioprobe_QC_SC[,-1], row.names = CTA_Bioprobe_QC_SC[,1])\n",
    "row.names(CTA_Bioprobe_QC_SC)[1:12] <- substring(row.names(CTA_Bioprobe_QC_SC)[1:12],1,12)\n",
    "row.names(CTA_Bioprobe_QC_SC)[13:18] <- substring(row.names(CTA_Bioprobe_QC_SC)[13:18],1,12)\n",
    "row.names(CTA_Bioprobe_QC_SC)[19:22] <- paste0(substring(row.names(CTA_Bioprobe_QC_SC)[19:22],1,9),\"P04\")\n",
    "row.names(CTA_Bioprobe_QC_SC)[23:24] <- substring(row.names(CTA_Bioprobe_QC_SC)[23:24],1,12)\n",
    "row.names(CTA_Bioprobe_QC_SC) <- gsub(\"Stroma\",\"CD45+\",rownames(CTA_Bioprobe_QC_SC))\n",
    "CTA_Bioprobe_QC_SC <- CTA_Bioprobe_QC_SC[order(row.names(CTA_Bioprobe_QC_SC)),]\n",
    "row.names(CTA_Bioprobe_QC_SC)[1:12] <- paste0(substr(row.names(CTA_Bioprobe_QC_SC)[1:12],1,5),substr(row.names(CTA_Bioprobe_QC_SC)[1:12],8,11),\n",
    "                                             substr(row.names(CTA_Bioprobe_QC_SC)[1:12],6,7))\n",
    "row.names(CTA_Bioprobe_QC_SC)[13:24] <- paste0(substr(row.names(CTA_Bioprobe_QC_SC)[13:24],1,6),substr(row.names(CTA_Bioprobe_QC_SC)[13:24],9,12),\n",
    "                                             substr(row.names(CTA_Bioprobe_QC_SC)[13:24],7,8))\n",
    "CTA_Bioprobe_QC_SC <- CTA_Bioprobe_QC_SC[order(row.names(CTA_Bioprobe_QC_SC),decreasing=T),]\n",
    "\n",
    "#################################################################################################################################################\n",
    "# Protein Data Wrangling\n",
    "#################################################################################################################################################\n",
    "\n",
    "Protein_QC$`Segment tags` <- paste(Protein_QC$`Segment tags`, Protein_QC$ROI_ID, Protein_QC$Scan_ID)\n",
    "Protein_QC_SC <- Protein_QC[,-c(1:6,8,9)] \n",
    "Protein_QC_SC$`Segment tags` <- gsub(\" \", \"_\", Protein_QC_SC$`Segment tags`)\n",
    "Protein_QC_SC <- as.data.frame(Protein_QC_SC)\n",
    "Protein_QC_SC <- data.frame(Protein_QC_SC[,-1], row.names = Protein_QC_SC[,1])\n",
    "row.names(Protein_QC_SC)[1:12] <- paste0(substring(row.names(Protein_QC_SC)[1:12],1,9),substring(row.names(Protein_QC_SC)[1:12],14,16))\n",
    "row.names(Protein_QC_SC)[13:18] <- paste0(substring(row.names(Protein_QC_SC)[13:18],1,9),substring(row.names(Protein_QC_SC)[13:18],14,16))\n",
    "row.names(Protein_QC_SC)[19:22] <- paste0(substring(row.names(Protein_QC_SC)[19:22],1,9),substring(row.names(Protein_QC_SC)[19:22],18,20))\n",
    "row.names(Protein_QC_SC)[23:24] <- paste0(substring(row.names(Protein_QC_SC)[23:24],1,9),substring(row.names(Protein_QC_SC)[23:24],14,16))\n",
    "row.names(Protein_QC_SC) <- gsub(\"Stroma\",\"CD45+\",rownames(Protein_QC_SC))\n",
    "Protein_QC_SC <- Protein_QC_SC[order(row.names(Protein_QC_SC)),]\n",
    "row.names(Protein_QC_SC)[1:12] <- paste0(substr(row.names(Protein_QC_SC)[1:12],1,5),substr(row.names(Protein_QC_SC)[1:12],8,11),\n",
    "                                              substr(row.names(Protein_QC_SC)[1:12],6,7))\n",
    "row.names(Protein_QC_SC)[13:24] <- paste0(substr(row.names(Protein_QC_SC)[13:24],1,6),substr(row.names(Protein_QC_SC)[13:24],9,12),\n",
    "                                               substr(row.names(Protein_QC_SC)[13:24],7,8))\n",
    "Protein_QC_SC <- Protein_QC_SC[order(row.names(Protein_QC_SC),decreasing=T),]\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "9555fe79",
   "metadata": {},
   "source": [
    "Processing Gene Expression Data"
   ]
  },
  {
   "cell_type": "code",
   "id": "f74f0edf",
   "metadata": {},
   "source": [
    "#################################################################################################################################################\n",
    "# EdgeR DEG\n",
    "#################################################################################################################################################\n",
    "\n",
    "groups <- factor(substring(row.names(CTA_Bioprobe_QC_SC),1,5))\n",
    "patients <- factor(rep(c(rep(\"RO1\",4),rep(\"P04\",2),rep(\"B18\",6)),2))\n",
    "\n",
    "CTA_Bioprobe_QC_SC = as.data.frame(t(CTA_Bioprobe_QC_SC))\n",
    "CTA_Bioprobe_QC_SC_DEG <- DGEList(CTA_Bioprobe_QC_SC, group=groups, genes= row.names(CTA_Bioprobe_QC_SC))\n",
    "CTA_Bioprobe_QC_SC_DEG$genes$Symbol <- row.names(CTA_Bioprobe_QC_SC_DEG)\n",
    "design <- model.matrix(~patients+groups)\n",
    "rownames(design) <- colnames(CTA_Bioprobe_QC_SC)\n",
    "keep <- filterByExpr(CTA_Bioprobe_QC_SC_DEG, design)\n",
    "CTA_Bioprobe_QC_SC_DEG <- CTA_Bioprobe_QC_SC_DEG[keep, , keep.lib.sizes=FALSE]\n",
    "AveLogCPM <- aveLogCPM(CTA_Bioprobe_QC_SC_DEG)\n",
    "CTA_Bioprobe_QC_SC_DEG <- calcNormFactors(CTA_Bioprobe_QC_SC_DEG)\n",
    "boxplot(log(CTA_Bioprobe_QC_SC), xlab=\"\", ylab=\"Log transformed raw counts\",las=2,main=\"Unnormalised logCPM\")\n",
    "\n",
    "colors <- c(\"red\", \"blue\", \"green\")\n",
    "pch <- c(15,16,17)\n",
    "plotMDS(CTA_Bioprobe_QC_SC_DEG,col=colors[patients],pch=pch[patients])\n",
    "legend(\"center\", legend=levels(patients), pch=pch, col=colors, ncol=1)\n",
    "\n",
    "plotMD(CTA_Bioprobe_QC_SC_DEG, column=1)\n",
    "abline(h=0, col=\"red\", lty=2, lwd=2)\n",
    "CTA_Bioprobe_QC_SC_DEG <- estimateDisp(CTA_Bioprobe_QC_SC_DEG, design, robust=TRUE)\n",
    "plotBCV(CTA_Bioprobe_QC_SC_DEG)\n",
    "\n",
    "fit <- glmQLFit(CTA_Bioprobe_QC_SC_DEG, design, robust=TRUE)\n",
    "plotQLDisp(fit)\n",
    "res_cta <- glmQLFTest(fit, coef=4)\n",
    "is.de <- decideTestsDGE(res_cta)\n",
    "plotMD(res_cta, status=is.de)\n",
    "tr <- glmTreat(fit, coef=4, lfc=log2(1))\n",
    "\n",
    "logcounts_cta <- edgeR::cpm(CTA_Bioprobe_QC_SC_DEG,log=TRUE)\n",
    "boxplot(logcounts_cta, xlab=\"\", ylab=\"Log2 counts per million\",las=2,main=\"Normalised logCPM\")\n",
    "logcounts <- logcounts_cta\n",
    "logcounts_cta_CD45 <- logcounts_cta[,1:12]\n",
    "logcounts_cta_PanCK <- logcounts_cta[,13:24]\n",
    "\n",
    "logcounts_cta_pat <- as.data.frame(logcounts_cta)\n",
    "my_sample_col <- data.frame(patient = patients)\n",
    "row.names(my_sample_col) <- colnames(logcounts_cta)\n",
    "my_sample_col <- as.data.frame(my_sample_col)\n",
    "my_sample_col$sample <- CTA_Bioprobe_QC_SC_DEG$samples$group\n",
    "annot_colors=list(sample=c(`CD45+`=\"#3A54A3\",PanCK=\"#3A783A\"),patient=c(B18=\"#F1A8B3\",\n",
    "                                                                        RO1=\"#91CFDA\",P04=\"#E96449\"))\n",
    "deg_tr <- as.data.frame(topTags(tr,n=Inf))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "d8d2a098",
   "metadata": {},
   "source": [
    "Processing Protein Expression Data"
   ]
  },
  {
   "cell_type": "code",
   "id": "ba15ca2e",
   "metadata": {},
   "source": [
    "#################################################################################################################################################\n",
    "# EdgeR Proteins\n",
    "#################################################################################################################################################\n",
    "\n",
    "groups_prot <- factor(substring(row.names(Protein_QC_SC),1,5))\n",
    "patients_prot <- patients\n",
    "Protein_QC_SC = as.data.frame(t(Protein_QC_SC))\n",
    "Protein_QC_SC_DEG <- DGEList(Protein_QC_SC, group=groups_prot, genes= row.names(Protein_QC_SC))\n",
    "Protein_QC_SC_DEG$genes$Symbol <- row.names(Protein_QC_SC_DEG)\n",
    "design_prot <- model.matrix(~patients_prot+groups_prot)\n",
    "rownames(design_prot) <- colnames(Protein_QC_SC)\n",
    "keep_prot <- filterByExpr(Protein_QC_SC_DEG, design_prot)\n",
    "Protein_QC_SC_DEG <- Protein_QC_SC_DEG[keep_prot, , keep.lib.sizes=FALSE]\n",
    "AveLogCPM <- aveLogCPM(Protein_QC_SC_DEG)\n",
    "Protein_QC_SC_DEG <- calcNormFactors(Protein_QC_SC_DEG)\n",
    "boxplot(log(Protein_QC_SC), xlab=\"\", ylab=\"Log transformed raw counts\",las=2,main=\"Unnormalised logCPM\")\n",
    "\n",
    "colors <- c(\"red\", \"blue\", \"green\")\n",
    "pch <- c(15,16,17)\n",
    "plotMDS(Protein_QC_SC_DEG,col=colors[patients_prot],pch=pch[patients_prot])\n",
    "legend(\"center\", legend=levels(patients_prot), pch=pch, col=colors, ncol=1)\n",
    "\n",
    "plotMD(Protein_QC_SC_DEG, column=1)\n",
    "abline(h=0, col=\"red\", lty=2, lwd=2)\n",
    "Protein_QC_SC_DEG <- estimateDisp(Protein_QC_SC_DEG, design_prot, robust=TRUE)\n",
    "plotBCV(Protein_QC_SC_DEG)\n",
    "\n",
    "fit_prot <- glmQLFit(Protein_QC_SC_DEG, design_prot, robust=TRUE)\n",
    "plotQLDisp(fit_prot)\n",
    "res_prot <- glmQLFTest(fit_prot, coef=4)\n",
    "is.de_prot <- decideTestsDGE(res_prot)\n",
    "plotMD(res_prot, status=is.de_prot)\n",
    "tr_prot <- glmTreat(fit_prot, coef=4, lfc=log2(1))\n",
    "\n",
    "logcounts_prot <- edgeR::cpm(Protein_QC_SC_DEG,log=TRUE)\n",
    "boxplot(logcounts_prot, xlab=\"\", ylab=\"Log2 counts per million\",las=2,main=\"Normalised logCPM\")\n",
    "logcounts_prot_CD45 <- logcounts_prot[,13:24]\n",
    "logcounts_prot_PanCK <- logcounts_prot[,1:12]\n",
    "\n",
    "logcounts_prot_pat <- as.data.frame(logcounts_prot)\n",
    "\n",
    "de_prot_tr <- as.data.frame(topTags(tr_prot,n=Inf))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "d2e06982",
   "metadata": {},
   "source": [
    "## Supplementary Fig10. A"
   ]
  },
  {
   "cell_type": "code",
   "id": "a146a9d9",
   "metadata": {},
   "source": [
    "pheatmap(logcounts_prot_pat[as.data.frame(filter(de_prot_tr,FDR<=0.05 & abs(logFC)>=1))$genes,],\n",
    "         scale=\"row\",annotation_col = my_sample_col,\n",
    "         cluster_cols = F, color=colorRampPalette(c(\"navy\", \"white\", \"red\"))(50), annotation_colors=annot_colors,\n",
    "         cellheight = 11, \n",
    "         cellwidth = 11, gaps_col=12)\n",
    "\n",
    "print(paste(\"The number of Differentially Expressed Proteins from a list of 48 proteins is:\",\n",
    "            length(as.data.frame(filter(de_prot_tr,FDR<=0.05 & abs(logFC)>=1))$genes)))\n",
    "\n",
    "#write.csv(as.data.frame(filter(de_prot_tr,FDR<=0.05 & abs(logFC)>=1)),\"NanoString_GeoMx_DSP_DEPs.csv\")\n",
    "#write.csv(topTags(tr_prot,n=Inf),\"NanoString_GeoMx_DSP_EdgeR_proteins_DEGScore.csv\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "749b49c2",
   "metadata": {},
   "source": [
    "## Supplementary Fig10. B"
   ]
  },
  {
   "cell_type": "code",
   "id": "fa1e8d08",
   "metadata": {},
   "source": [
    "pheatmap(as.matrix(as.data.frame(logcounts_cta)[c(\"CD163\",\"CD68\",\"MS4A1\",\"CD8A\",\"CD8B\",\"ITGAX\",\"FOXP3\",\"IL2RA\",\"FAP\",\"FN1\"),]),\n",
    "         annotation_col = my_sample_col, scale = \"row\",fontsize = 10, cellheight = 9, cellwidth = 9,\n",
    "         cluster_rows=F, cluster_cols = F, gaps_col = c(12), color=colorRampPalette(c(\"navy\", \"white\", \"red\"))(100),\n",
    "         annotation_colors = annot_colors)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "5b30f13b",
   "metadata": {},
   "source": [
    "## Supplementary Fig10. C"
   ]
  },
  {
   "cell_type": "code",
   "id": "8f0cafcf",
   "metadata": {},
   "source": [
    "pheatmap(logcounts_cta_pat[filter(deg_tr,FDR<=0.05 & abs(logFC)>=1.0)$genes[1:30],],\n",
    "         scale=\"row\",annotation_col = my_sample_col,\n",
    "         cluster_cols = F, color=colorRampPalette(c(\"navy\", \"white\", \"red\"))(50), annotation_colors=annot_colors,\n",
    "         cellheight = 11, \n",
    "         cellwidth = 11, gaps_col=12)\n",
    "\n",
    "\n",
    "print(paste(\"The number of Differentially Expressed Proteins from a list of 1820 genes is:\",\n",
    "            length(as.data.frame(filter(deg_tr,FDR<=0.05 & abs(logFC)>=1.0))$genes)))\n",
    "\n",
    "#write.csv(as.data.frame(filter(deg_tr,FDR<=0.05 & abs(logFC)>=1.0)),\"NanoString_GeoMx_DSP_DEGs.csv\")\n",
    "#write.csv(topTags(tr,n=Inf),\"NanoString_GeoMx_DSP_EdgeR_genelist_DEGScore.csv\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "faf4b6d0",
   "metadata": {},
   "source": [
    "## Supplementary Fig10. D"
   ]
  },
  {
   "cell_type": "code",
   "id": "d4122a0d",
   "metadata": {},
   "source": [
    "# No proteins overlapped\n",
    "DE_proteins <- as.data.frame(topTags(res_prot,n=34))\n",
    "DE_proteins <- DE_proteins[c(all_cancer_PatientOverlapResults$genes,all_normal_PatientOverlapResults$genes),]\n",
    "DE_proteins <- na.omit(DE_proteins)\n",
    "\n",
    "\n",
    "DE_genes <- as.data.frame(filter(deg_tr,FDR<=0.05 & abs(logFC)>=1.0))\n",
    "DE_genes <- DE_genes[c(all_cancer_PatientOverlapResults$genes,all_normal_PatientOverlapResults$genes),]\n",
    "DE_genes <- na.omit(DE_genes)\n",
    "DE_genes$group <- c(rep(\"Cancer\",6),rep(\"Normal\",27))\n",
    "my_sample_row <- DE_genes[c(\"group\")]\n",
    "\n",
    "pheatmap(as.matrix(as.data.frame(logcounts_cta)[DE_genes$genes,]), \n",
    "         annotation_col = my_sample_col, annotation_row = my_sample_row,\n",
    "         cutree_cols = 2, scale = \"row\",fontsize = 10, cellheight = 9, cellwidth = 9,\n",
    "         cluster_rows = FALSE, color=colorRampPalette(c(\"navy\", \"white\", \"red\"))(15), annotation_colors=annot_colors,\n",
    "         cluster_cols = F, gaps_col=12)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "d971d857",
   "metadata": {},
   "source": [
    "## Supplementary Fig10. E"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62d25b6a",
   "metadata": {},
   "source": [
    "https://github.com/BiomedicalMachineLearning/scc_multiomics/tree/main/Development/SkinCancerAtlasInteractome/FigS4_NanoString/FigS4c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7409fbec",
   "metadata": {},
   "source": [
    "## Supplementary Fig10. F"
   ]
  },
  {
   "cell_type": "code",
   "id": "4fda8438",
   "metadata": {},
   "source": [
    "#################################################################################################################################################\n",
    "# Cell De-convolution\n",
    "#################################################################################################################################################\n",
    "\n",
    "CTA_Bioprobe_QC_SC_norm <- data.matrix(edgeR::cpm(CTA_Bioprobe_QC_SC,log = TRUE))\n",
    "\n",
    "per.observation.mean.neg = CTA_Bioprobe_QC_SC_norm[\"NegProbe\", ]\n",
    "bg = sweep(CTA_Bioprobe_QC_SC_norm * 0, 2, per.observation.mean.neg, \"+\")\n",
    "heatmap(sweep(safeTME, 1, apply(safeTME, 1, max), \"/\"),\n",
    "        labRow = NA, margins = c(10, 5))\n",
    "\n",
    "bg2 = derive_GeoMx_background(norm = CTA_Bioprobe_QC_SC_norm,\n",
    "                              probepool = rep(1, nrow(CTA_Bioprobe_QC_SC_norm)),\n",
    "                              negnames = \"NegProbe\")\n",
    "\n",
    "res = spatialdecon(norm = CTA_Bioprobe_QC_SC_norm,\n",
    "                   bg = bg,\n",
    "                   X = safeTME,\n",
    "                   align_genes = TRUE)\n",
    "\n",
    "xm <- melt(res$beta, id.vars = \"Cell-Type\", variable.name=\"Samples\", value.name = \"Size\")\n",
    "\n",
    "# Advanced\n",
    "Segment_Properties <- Segment_Properties[order(Segment_Properties$SegmentLabel),]\n",
    "Segment_Properties$is_PanCK = (Segment_Properties$SegmentLabel==\"PanCK+\")\n",
    "restils = spatialdecon(norm = CTA_Bioprobe_QC_SC_norm,                     \n",
    "                       raw = data.matrix(CTA_Bioprobe_QC_SC),                        \n",
    "                       bg = bg, X = safeTME,                     \n",
    "                       cellmerges = safeTME.matches,   \n",
    "                       cell_counts = c(864,431,157,137,289,332,66,47,93,227,49,119,330,\n",
    "                                       416,464,221,516,1040,273,423,352,1165,878,1103),      \n",
    "                       is_pure_tumor = Segment_Properties$SegmentLabel==\"PanCK+\",   \n",
    "                       n_tumor_clusters = 12) \n",
    "\n",
    "#################################################################################################################################################\n",
    "# Cell De-convolution\n",
    "#################################################################################################################################################\n",
    "\n",
    "coul3 <- c(\"deeppink\", \"gold\", \"navyblue\", \"orange\", \"maroon\", \"chartreuse4\", \"darkorchid2\",\n",
    "           \"yellowgreen\", \"darkmagenta\", \"lightseagreen\",\"gold4\", \"orangered\", \"limegreen\", \"magenta\")\n",
    "order_cell_type <- c(\"CD4.T.cells\", \"macrophages\", \"Treg\", \"endothelial.cells\", \"B\", \"fibroblasts\", \"NK\",\n",
    "            \"CD8.T.cells\", \"mast\", \"pDC\", \"plasma\", \"monocytes\", \"mDCs\", \"neutrophils\")\n",
    "\n",
    "cell_type_proportion <-  restils$prop_of_all \n",
    "data_percentage_cell_type_proportion <- apply(cell_type_proportion, 2, function(x){x*100/sum(x,na.rm=T)})\n",
    "data_percentage_cell_type_proportion_stroma <- data_percentage_cell_type_proportion[,13:24] \n",
    "data_percentage_cell_type_proportion_stroma <- data.frame(data_percentage_cell_type_proportion_stroma) %>%slice(match(order_cell_type, \n",
    "                                                                                                                    rownames(data_percentage_cell_type_proportion_stroma)))\n",
    "rownames(data_percentage_cell_type_proportion_stroma) <- c(\"CD4 T cells\", \"macrophages\", \"Treg cells\", \"endothelial cells\", \"B cells\", \"fibroblasts\", \"NK cells\",\n",
    "                                                          \"CD8 T cells\", \"mast cells\", \"pDC\", \"plasma\", \"monocytes\", \"mDCs\", \"neutrophils\")\n",
    "colnames(data_percentage_cell_type_proportion_stroma) <- substr(colnames(data_percentage_cell_type_proportion_stroma),7,22)\n",
    "bar <- barplot(as.matrix(data_percentage_cell_type_proportion_stroma), col = coul3, border=\"white\", cex.names=1.0, las=2, \n",
    "               width=c(330,416,464,221,516,1040,273,423,352,1165,878,1103), space = c(0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5))\n",
    "legend <- barplot(as.matrix(data_percentage_cell_type_proportion_stroma), col = coul3, legend=TRUE)\n",
    "\n",
    "#write.csv(data_percentage_cell_type_proportion_stroma,\"NanoString_GeoMx_DSP_Cell_type_proportions.csv\") \n",
    "#save(restils, file=\"NanoString_GeoMx_DSP_Cell_type_proportions_object.Rdata\")"
   ],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
