{
 "cells": [
  {
   "cell_type": "code",
   "id": "97b533df-fd75-42b5-baee-1ba22ecee2c7",
   "metadata": {},
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from tqdm.auto import tqdm\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import adjusted_rand_score\n",
    "from sklearn.cluster import AgglomerativeClustering"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "223cdec1-3b3f-4d0d-af82-2f93308d970c",
   "metadata": {},
   "source": [
    "import sys\n",
    "print(sys.executable)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "29e8a9f4-f094-447b-919e-f460b1035c52",
   "metadata": {},
   "source": [
    "! pip show leidenalg"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "0741953c-7769-42f9-bae6-b84e281742fb",
   "metadata": {},
   "source": [
    "#python -m pip install stlearn igraph"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "91190f29-35d7-473c-8be3-af2a5f45b895",
   "metadata": {},
   "source": [
    "## Reading the count data and metadata files\n",
    "batch = pd.read_csv(\"/QRISdata/Q1851/Prakrithi/Melanoma_scRNAseq/MPSs_count_metadata.txt\", sep=\"\\t\", header=0, index_col='BC')\n",
    "#batch=batch.set_index('BC')\n",
    "rna_counts = np.transpose(pd.read_csv(\"/QRISdata/Q1851/Prakrithi/Melanoma_scRNAseq/MPSs_count_mat.txt\", sep=\"\\t\", header=0,index_col=0))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ead55e58-1d1a-4e0a-a081-1110babc7479",
   "metadata": {},
   "source": [
    "## creating anndata objects from counts\n",
    "rna_adata = sc.AnnData(rna_counts)\n",
    "rna_adata.obs=batch.iloc[:,[0,1,2]]\n",
    "## Adding batch information to the Anndata objects from the metadata file\n",
    "#rna_adata.obs['sample'] = batch['orig.ident']\n",
    "## Making rows and column names unique (Cell barcode information is found in .obs and gene (feature) information in .var)\n",
    "rna_adata.var_names_make_unique()\n",
    "rna_adata.obs_names_make_unique()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "25a9c936-77a2-4187-a77c-b1bfb382b052",
   "metadata": {},
   "source": [
    "rna_adata"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bff27cd1-ca1f-4161-a2b3-152874a751dc",
   "metadata": {},
   "source": [
    "#! pip install -U scikit-image on terminal\n",
    "import skimage"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "a3eeb66e-e778-4b32-9a37-a560f5a8faf4",
   "metadata": {},
   "source": [
    "#Doublet testing\n",
    "sc.pp.scrublet(rna_adata, batch_key=\"orig.ident\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5eafc318-abc0-4dda-8063-eacb3163086b",
   "metadata": {},
   "source": [
    "#rna_adata.obs\n",
    "\n",
    "rna_adata.obs[rna_adata.obs['predicted_doublet'] == True]\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "0f02c80f-20cc-4ce4-a9a0-f895332252c0",
   "metadata": {},
   "source": [
    "rna_adata.obs.to_csv('/QRISdata/Q1851/Prakrithi/Melanoma_scRNAseq/rna_adata_obs.csv', sep=\"\\t\",index=True)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4c1a9d44-5ff3-4ebd-8105-add9c31d9e7e",
   "metadata": {},
   "source": [
    "sc.pl.violin(rna_adata, ['doublet_score'])\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4f2b24b5-0e29-4425-b64c-1ead53e5d4f4",
   "metadata": {},
   "source": [
    "## QC plots for total genes, counts and percentage of mitochondrial genes in cells\n",
    "rna_adata.var['mt'] = rna_adata.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(rna_adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "# Scanpy will prepend the string in the save argument with \"violin\"\n",
    "# and save it to our figure directory defined in the first step.\n",
    "sc.pl.violin(rna_adata, ['n_genes_by_counts'], save='_n_genes', jitter=0.4)\n",
    "sc.pl.violin(rna_adata, ['total_counts'], save='_total_counts', jitter=0.4)\n",
    "sc.pl.violin(rna_adata, ['pct_counts_mt'], save='_mito_pct', jitter=0.4)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8aee631d-a776-436d-a955-15ee3cce1abe",
   "metadata": {},
   "source": [
    "# Filtering out the low quality genes and cells\n",
    "rna_adata = rna_adata[(rna_adata.obs.n_genes_by_counts < 7000) &\n",
    "(rna_adata.obs.total_counts < 20000),:]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "2b8d3bc8-3407-4025-8919-8d4636622eeb",
   "metadata": {},
   "source": [
    "## Normalizing the counts\n",
    "sc.pp.normalize_total(rna_adata, target_sum=1e4)\n",
    "sc.pp.log1p(rna_adata)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "3b76f4dc-daeb-4644-9e73-025a5b8452bb",
   "metadata": {},
   "source": [
    "## Identifying the highly variable genes\n",
    "sc.pp.highly_variable_genes(rna_adata, min_mean=0.0125, max_mean=3, min_disp=0.25)\n",
    "# This saves the original set of genes\n",
    "rna_adata.raw = rna_adata\n",
    "\n",
    "rna_adata = rna_adata[:,rna_adata.var.highly_variable]\n",
    "sc.pp.scale(rna_adata, max_value=10)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "063332af-9ba4-48c1-a089-d9a3ca656bf1",
   "metadata": {},
   "source": [
    "# Lower dimension embedding - PCA\n",
    "sc.tl.pca(rna_adata, svd_solver='arpack',n_comps=200)\n",
    "sc.pl.pca_variance_ratio(rna_adata, log=True, n_pcs=200, save='200pc') # scanpy generates the filename automatically"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "3f62f261-e883-4c97-a70a-91fd5e620a58",
   "metadata": {},
   "source": [
    "# Lower dimension embedding - UMAP and clustering\n",
    "sc.pp.neighbors(rna_adata, n_neighbors=30, n_pcs=75)\n",
    "sc.tl.umap(rna_adata)\n",
    "sc.tl.leiden(rna_adata, resolution=0.4)\n",
    "sc.pl.umap(rna_adata, color=['leiden'] , legend_loc = 'best')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5f1c19f2-01b1-4fa4-99a1-a3e85e36374c",
   "metadata": {},
   "source": [
    "sc.tl.leiden(rna_adata, resolution=0.4)\n",
    "sc.pl.umap(rna_adata, color=['leiden'] , legend_loc = 'best')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "896ec895-1c5a-46fb-ac43-923f0f4fad57",
   "metadata": {},
   "source": [
    "def one_col_lgd(umap):\n",
    "    legend = umap.legend(bbox_to_anchor=[1.00, 0.5],\n",
    "    loc='center left', ncol=1, prop={'size': 6})\n",
    "    legend.get_frame().set_linewidth(0.0)\n",
    "    for handle in legend.legendHandles:\n",
    "        handle.set_sizes([25.0])\n",
    "    return legend"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5730e661-4329-41df-9ac0-c12e734b513b",
   "metadata": {},
   "source": [
    "## Clustering wiithout batch correction\n",
    "donor_umap = sc.pl.umap(rna_adata, color=['orig.ident'],\n",
    "show=False, palette=sns.color_palette(\"husl\", 24),\n",
    "    legend_fontsize=6, frameon=True, title='Donor')\n",
    "\n",
    "lgd = one_col_lgd(donor_umap)\n",
    "\n",
    "fig = donor_umap.get_figure()\n",
    "fig.set_size_inches(5, 5)\n",
    "fig.savefig('/QRISdata/Q1851/Prakrithi/Melanoma_scRNAseq/umap_lgd_sample75pc_n',\n",
    "    dpi=400, bbox_extra_artists=(lgd,), bbox_inches='tight')\n",
    "\n",
    "# by cluster\n",
    "leiden_umap = sc.pl.umap(rna_adata, color=['leiden'],\n",
    "    show=False, palette=sns.color_palette(\"husl\", 24),\n",
    "legend_fontsize=6, frameon=True, title='Leiden')\n",
    "\n",
    "lgd = one_col_lgd(leiden_umap)\n",
    "\n",
    "fig = leiden_umap.get_figure()\n",
    "fig.set_size_inches(5, 5)\n",
    "fig.savefig('/QRISdata/Q1851/Prakrithi/Melanoma_scRNAseq/umap_lgd_leiden75pc_n',\n",
    "    dpi=400, bbox_extra_artists=(lgd,), bbox_inches='tight')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9da3a28e-ec5b-4598-8864-b2a3c5f9d9ba",
   "metadata": {},
   "source": [
    "## Batch correction\n",
    "sc.external.pp.harmony_integrate(rna_adata, 'orig.ident')\n",
    "rna_adata.obsm['X_pca'] = rna_adata.obsm['X_pca_harmony']\n",
    "sc.pp.neighbors(rna_adata, n_neighbors=30, n_pcs=75)\n",
    "sc.tl.umap(rna_adata)\n",
    "sc.tl.leiden(rna_adata, resolution=0.4)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5da930dc-b1d6-439b-aec1-2f60e3549136",
   "metadata": {},
   "source": [
    "### Clustering after batch correction\n",
    "# by sample\n",
    "donor_umap = sc.pl.umap(rna_adata, color=['orig.ident'],\n",
    "    show=False, legend_fontsize=6, frameon=True, title='Donor')\n",
    "\n",
    "lgd = one_col_lgd(donor_umap)\n",
    "\n",
    "fig = donor_umap.get_figure()\n",
    "fig.set_size_inches(5, 5)\n",
    "fig.savefig('/QRISdata/Q1851/Prakrithi/Melanoma_scRNAseq/umap_lgd_harmony_sample75pcs_n',\n",
    "    dpi=400, bbox_extra_artists=(lgd,), bbox_inches='tight')\n",
    "\n",
    "# by cluster\n",
    "leiden_umap = sc.pl.umap(rna_adata, color=['leiden'],\n",
    "show=False, palette=sns.color_palette(\"husl\", 24),\n",
    "    legend_fontsize=6, frameon=True, title='Leiden')\n",
    "\n",
    "lgd = one_col_lgd(leiden_umap)\n",
    "\n",
    "fig = leiden_umap.get_figure()\n",
    "fig.set_size_inches(5, 5)\n",
    "fig.savefig('/QRISdata/Q1851/Prakrithi/Melanoma_scRNAseq/umap_lgd_harmony_leiden75pcs_n',\n",
    "    dpi=400, bbox_extra_artists=(lgd,), bbox_inches='tight')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ff5a8200-af60-4f44-96fd-31007939323a",
   "metadata": {},
   "source": [
    "sc.pl.umap(rna_adata, color=\"leiden\", legend_loc='on data')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "3bcec558-779e-4397-8bab-837f4fee1596",
   "metadata": {},
   "source": [
    "sc.pl.umap(rna_adata, color=['doublet_score'] , legend_loc = 'best')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "399055df-5b18-4f70-b3e6-c25603e945b8",
   "metadata": {},
   "source": [
    "rna_adata"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4f0b46d4-c494-4ad8-8e9b-36c27d427f64",
   "metadata": {},
   "source": [
    "sc.pl.umap(rna_adata, color=\"leiden\", legend_loc='on data')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e5fe2e8e-82e3-4732-97e7-054b91b07852",
   "metadata": {},
   "source": [
    "sc.tl.leiden(rna_adata, key_added=\"leiden_res0_25\", resolution=0.25)\n",
    "sc.tl.leiden(rna_adata, key_added=\"leiden_res0_5\", resolution=0.5)\n",
    "sc.tl.leiden(rna_adata, key_added=\"leiden_res1\", resolution=1.0)\n",
    "sc.pl.umap(\n",
    "    rna_adata,\n",
    "    color=[\"leiden_res0_25\", \"leiden_res0_5\", \"leiden_res1\"],\n",
    "    legend_loc=\"on data\",\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b9acd435-91ba-44ab-bcd0-e8e2973047a5",
   "metadata": {},
   "source": [
    "rna_adata.raw = rna_adata\n",
    "rna_adata = rna_adata[:, rna_adata.var[\"highly_variable\"]].copy()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "f4d0b53a-48b5-498c-8563-144fbbcd9081",
   "metadata": {},
   "source": [
    "sc.tl.rank_genes_groups(rna_adata, groupby=\"leiden\", method=\"wilcoxon\")\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    rna_adata, groupby=\"leiden\", standard_scale=\"var\", n_genes=5\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "1e697c67-d844-4c91-aa9a-7fc90e664fa6",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8ad94d18-9566-4145-9a16-aa98be6ada8c",
   "metadata": {},
   "source": [
    "# 6, 14 are doublets with keratinocytes?? Express KRT14\n",
    "\n",
    "new_cell_dict = {\n",
    "    \"0\": \"Melanocytes\",\n",
    "    \"1\": \"KC\", #\n",
    "    \"2\": \"Fibroblast\",\n",
    "    \"3\": \"Immune\", #T\n",
    "    \"4\": \"KC\", #Basal\n",
    "    \"5\": \"Immune\", #DC\n",
    "    \"6\": \"KC\", #\n",
    "    \"7\": \"Immune\", #T\n",
    "    \"8\": \"Endothelial cell\", #\n",
    "    \"9\": \"KC\", #\n",
    "    \"10\": \"KC\", #\n",
    "    \"11\": \"KC\", #\n",
    "    \"12\": \"Melanocytes\", #\n",
    "    \"13\": \"KC\", #Basal\n",
    "    \"14\": \"Fibroblast\",\n",
    "    \"15\": \"Immune\", #DC\n",
    "    \"16\": \"Immune\", #DC\n",
    "    \"17\": \"Immune\", #NK\n",
    "    \"18\": \"Immune\", #LC (Immature DC)\n",
    "    \"19\": \"Endothelial cell\",\n",
    "    \"20\": \"Immune\", #DC\n",
    "    \"21\": \"KC\", #stem/hair/stress\n",
    "    \"22\": \"\"\n",
    "}\n",
    "\n",
    "new_cell_dict2 = {\n",
    "    \"0\": \"Melanocytes\",\n",
    "    \"1\": \"KC Differentiating\",\n",
    "    \"2\": \"Fibroblast\",\n",
    "    \"3\": \"Treg\",\n",
    "    \"4\": \"KC Basal\",\n",
    "    \"5\": \"Macrophage\",\n",
    "    \"6\": \"KC Cornified\", #?\n",
    "    \"7\": \"T cell\",\n",
    "    \"8\": \"Endothelial cell\",\n",
    "    \"9\": \"KC Differentiating\", #?\n",
    "    \"10\": \"KC nail\",\n",
    "    \"11\": \"KC stem/hair/stress\",\n",
    "    \"12\": \"Melanocytes\",\n",
    "    \"13\": \"KC Basal\",\n",
    "    \"14\": \"Fibroblast\",\n",
    "    \"15\": \"DC\",  #?\n",
    "    \"16\": \"DC\", #pDC\n",
    "    \"17\": \"NK\",\n",
    "    \"18\": \"LC\",\n",
    "    \"19\": \"Endothelial cell\",\n",
    "    \"20\": \"DC\", #?\n",
    "    \"21\": \"KC stem/hair/stress\", \n",
    "    \"22\": \"\"\n",
    "}\n",
    "\n",
    "new_cell_dict2 = {\n",
    "    \"0\": \"Melanocytes\",\n",
    "    \"1\": \"KC Differentiating\",\n",
    "    \"2\": \"Fibroblast\",\n",
    "    \"3\": \"Treg\",\n",
    "    \"4\": \"KC Basal\",\n",
    "    \"5\": \"Macrophage\",\n",
    "    \"6\": \"KC Cornified\", #?\n",
    "    \"7\": \"T cell\",\n",
    "    \"8\": \"Endothelial cell\",\n",
    "    \"9\": \"KC Differentiating\", #?\n",
    "    \"10\": \"KC nail\",\n",
    "    \"11\": \"KC stem/hair/stress\",\n",
    "    \"12\": \"Melanocytes\",\n",
    "    \"13\": \"KC Basal\",\n",
    "    \"14\": \"Fibroblast\",\n",
    "    \"15\": \"DC\",  #?\n",
    "    \"16\": \"DC\", #pDC\n",
    "    \"17\": \"NK\",\n",
    "    \"18\": \"LC\",\n",
    "    \"19\": \"Endothelial cell\",\n",
    "    \"20\": \"DC\", #?\n",
    "    \"21\": \"KC stem/hair/stress\", \n",
    "    \"22\": \"\"\n",
    "}\n",
    "\n",
    "rna_adata.obs[\"Level1\"] = [new_cell_dict[x] for x in rna_adata.obs[\"leiden\"]]\n",
    "rna_adata.obs[\"Level2\"] = [new_cell_dict2[x] for x in rna_adata.obs[\"leiden\"]]\n",
    "\n",
    "#mdata = idata[idata.obs[\"Level1\"].isin([\"Imm_Myeloid\"])].copy()\n",
    "#tdata = idata[idata.obs[\"Level1\"].isin([\"Imm_T cell\"])].copy()\n",
    "#mdata = idata[idata.obs[\"Level1\"].isin([\"Imm_Myeloid\"])].copy()\n",
    "#tdata = idata[idata.obs[\"Level1\"].isin([\"Imm_T cell\"])].copy()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "69c0afc7-78f4-412e-ba8c-ff32fa9e5622",
   "metadata": {},
   "source": [
    "sc.pl.umap(\n",
    "    rna_adata,\n",
    "    color=[\"Level1\",\"Level2\"],\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "454085ad-9cec-4760-8556-e458ea2a8a99",
   "metadata": {},
   "source": [
    "#from tools import cmp\n",
    "\n",
    "sc.pl.umap(\n",
    "    rna_adata,\n",
    "    color=[\"CD3E\",\"KRT10\",\"KRT6A\",\n",
    "        \"KRT17\",\n",
    "        \"KRT15\",\n",
    "        \"KRT2\",\n",
    "        \"LOR\",\n",
    "        \"IVL\",\n",
    "    ]\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b626ab79-d6be-4841-aa27-201d908c94d8",
   "metadata": {},
   "source": [
    "sc.pl.umap(\n",
    "    rna_adata,\n",
    "    color=[\n",
    "        \"RRM2\",\n",
    "        \"HELLS\",\n",
    "        \"UHRF1\",\n",
    "        \"ASS1\",\n",
    "        \"COL17A1\",\n",
    "        \"POSTN\",\n",
    "        \"KRT19\",\n",
    "        \"GJB2\",\n",
    "        \"KRT6A\",\n",
    "        \"KRT16\",\n",
    "        \"CCND1\",\n",
    "        \"DEFB1\",\n",
    "        \"CALML3\",\n",
    "        \"CALML5\",\n",
    "        \"ZNF750\",\n",
    "        \"SPINK5\",\n",
    "        \"CAPN3\",\n",
    "    ]\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ac064fda-2c09-4150-8205-1a6a55b4a2fe",
   "metadata": {},
   "source": [
    "# Immune subclusters\n",
    "sc.pl.dotplot(\n",
    "    rna_adata,\n",
    "    {\n",
    "        \"General\": [\"ITGAM\"],\n",
    "        \"Melanocyte\": [\"MITF\",\"PMEL\",\"TYR\",\"DCT\",\"MLANA\",\"PMEL\",\"S100A1\"], \n",
    "        \"LC\": [\"CD207\", \"EPCAM\", \"ITGAX\"],\n",
    "        \"Monocyte\": [\"CCR2\", \"CD14\", \"S100A8\", \"S100A9\", \"CX3CR1\"],\n",
    "        \"DC1\": [\"XCR1\", \"CLEC9A\"],\n",
    "        \"DC2\": [\"CLEC10A\", \"THBD\"],\n",
    "        \"DC3\": [\"CSF3R\", \"CSF2RA\"],\n",
    "        \"mRegDC\": [\"LAMP3\", \"CCR7\"],\n",
    "        \"pDC\": [\"IL3RA\", \"KIT\", \"IRF8\"],\n",
    "        \"Mac\": [\n",
    "            \"C1QA\",\n",
    "            \"TREM2\",\n",
    "            \"SIGLEC1\",\n",
    "            \"CD68\",\n",
    "            \"LYVE1\",\n",
    "            \"ARG1\",\n",
    "            \"CD163\",\n",
    "        ],\n",
    "        \"prolif\": [\"UBE2C\",\"NUSAP1\",\"MKI67\"],\n",
    "        \"IFN\": [\"ISG15\", \"IFI27\", \"STAT1\"],\n",
    "        \"contamination\": [\"CD3D\"],\n",
    "    },\n",
    "    standard_scale=\"var\",\n",
    "    color_map=\"Blues\",\n",
    "    groupby=\"leiden\",\n",
    ") #\"leiden_res0_25\""
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "27f02317-3d61-4df7-ae6c-231b36c275bb",
   "metadata": {},
   "source": [
    "# Not as useful compared to the umap gene expression graphs\n",
    "\n",
    "sc.pl.dotplot(\n",
    "    rna_adata,\n",
    "    {\n",
    "        \"Immune cell\": [\n",
    "            \"CD3D\",\n",
    "            \"CD3E\",\n",
    "            \"CD4\",\n",
    "            \"CD8A\",\n",
    "            \"CCR7\",\n",
    "            \"SELL\",\n",
    "            \"CD27\",\n",
    "            \"ITGAM\",\n",
    "            \"S100A8\",\n",
    "        ],\n",
    "        \"prolif\": [\"MKI67\"],\n",
    "        \"IFN\": [\"ISG15\", \"IFI27\", \"STAT1\"],\n",
    "        \"KC Basal\": [\"KRT15\", \"C1orf56\"],\n",
    "        \"KC Differentiating\": [\"KRT10\"],\n",
    "        \"KC Cornified\": [\"LOR\", \"IVL\"],\n",
    "        \"KC Granular\": [\"KRT2\"],\n",
    "        \"KC stem/hair/stress\": [\"KRT16\", \"KRT17\", \"KRT6A\"],\n",
    "        \"KC nail/nail\": [\n",
    "            \"KRT6B\",\n",
    "            \"KRT79\",\n",
    "        ],\n",
    "    },\n",
    "    standard_scale=\"var\",\n",
    "    color_map=\"Blues\",\n",
    "    groupby=\"leiden\",\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b93dbf7d-3cc4-4bd3-a72b-acdcc0c28ed5",
   "metadata": {},
   "source": [
    "sc.pl.dotplot(\n",
    "    rna_adata,\n",
    "    {\n",
    "        \"T cell\": [\n",
    "            \"CD3D\",\n",
    "            \"CD3E\",\n",
    "            \"CD4\",\n",
    "            \"CD8A\",\n",
    "            \"CCR7\",\n",
    "            \"SELL\",\n",
    "            \"CD27\",\n",
    "        ],\n",
    "        \"Trm\": [\"CD69\", \"CRTAM\", \"S1PR1\"],\n",
    "        \"Treg\": [\"FOXP3\", \"PDCD1\", \"CTLA4\", \"IL2RA\"],\n",
    "        \"NK\": [ \"NCR1\", \"GZMB\",\"GNLY\", \"XCL2\"],\n",
    "        \"ILC\": [\"IL7R\"],\n",
    "        \"prolif\": [\"MKI67\"],\n",
    "        \"IFN\": [\"ISG15\", \"IFI27\", \"STAT1\"],\n",
    "        \"Contamination\": [\"ITGAM\", \"S100A8\"],\n",
    "        \"Endothelial\": [\"CLDN5\", \"PLVAP\", \"SPARCL1\"],\n",
    "        \"Fibroblast\": [\"PDGFRA\", \"PDGFRB\", \"DCN\"]\n",
    "    },\n",
    "    standard_scale=\"var\",\n",
    "    color_map=\"Blues\",\n",
    "    groupby=\"leiden\",\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "561b5606-ab48-4273-8d62-7a35f20fae54",
   "metadata": {},
   "source": [
    "# Cluster verification"
   ]
  },
  {
   "cell_type": "code",
   "id": "6b5f7a16-dd69-471a-8c14-56124d396fff",
   "metadata": {},
   "source": [
    "sc.pl.dotplot(\n",
    "    rna_adata,\n",
    "    {\n",
    "        \"Immune cell\": [\n",
    "            \"CD3D\",\n",
    "            \"CD3E\",\n",
    "            \"CD4\",\n",
    "            \"CD8A\",\n",
    "            \"CCR7\",\n",
    "            \"SELL\",\n",
    "            \"CD27\",\n",
    "            \"ITGAM\",\n",
    "            \"S100A8\",\n",
    "        ],\n",
    "        \"prolif\": [\"MKI67\"],\n",
    "        \"IFN\": [\"ISG15\", \"IFI27\", \"STAT1\"],\n",
    "        \"KC Basal\": [\"KRT15\", \"C1orf56\"],\n",
    "        \"KC Differentiating\": [\"KRT10\"],\n",
    "        \"KC Cornified\": [\"LOR\", \"IVL\"],\n",
    "        \"KC Granular\": [\"KRT2\"],\n",
    "        \"KC stem/hair/stress\": [\"KRT16\", \"KRT17\", \"KRT6A\"],\n",
    "        \"KC nail/nail\": [\n",
    "            \"KRT6B\",\n",
    "            \"KRT79\",\n",
    "        ],\n",
    "               \"General\": [\"ITGAM\"],\n",
    "        \"Melanocyte\": [\"MITF\",\"PMEL\",\"TYR\",\"DCT\",\"MLANA\",\"PMEL\",\"S100A1\"], \n",
    "        \"LC\": [\"CD207\", \"EPCAM\", \"ITGAX\"],\n",
    "        \"Monocyte\": [\"CCR2\", \"CD14\", \"S100A8\", \"S100A9\", \"CX3CR1\"],\n",
    "        \"DC1\": [\"XCR1\", \"CLEC9A\"],\n",
    "        \"DC2\": [\"CLEC10A\", \"THBD\"],\n",
    "        \"DC3\": [\"CSF3R\", \"CSF2RA\"],\n",
    "        \"mRegDC\": [\"LAMP3\", \"CCR7\"],\n",
    "        \"pDC\": [\"IL3RA\", \"KIT\", \"IRF8\"],\n",
    "        \"Mac\": [\n",
    "            \"C1QA\",\n",
    "            \"TREM2\",\n",
    "            \"SIGLEC1\",\n",
    "            \"CD68\",\n",
    "            \"LYVE1\",\n",
    "            \"ARG1\",\n",
    "            \"CD163\",\n",
    "        ],\n",
    "        \"Endothelial\": [\"CLDN5\", \"PLVAP\", \"SPARCL1\"],\n",
    "        \"Fibroblast\": [\"PDGFRA\", \"PDGFRB\", \"DCN\"]\n",
    "    },\n",
    "    standard_scale=\"var\",\n",
    "    color_map=\"Blues\",\n",
    "    groupby=\"Level2\",\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "0637de8d-b561-45f7-8a77-7bc0ec73d017",
   "metadata": {},
   "source": [
    "## res1 clusters"
   ]
  },
  {
   "cell_type": "code",
   "id": "2486c3a3-8587-455d-a336-2a4665ad63cd",
   "metadata": {},
   "source": [
    "# Immune subclusters\n",
    "sc.pl.dotplot(\n",
    "    rna_adata,\n",
    "    {\n",
    "        \"General\": [\"ITGAM\"],\n",
    "        \"Melanocyte\": [\"MITF\",\"PMEL\",\"TYR\",\"DCT\",\"MLANA\",\"PMEL\",\"S100A1\"], \n",
    "        \"LC\": [\"CD207\", \"EPCAM\", \"ITGAX\"],\n",
    "        \"Monocyte\": [\"CCR2\", \"CD14\", \"S100A8\", \"S100A9\", \"CX3CR1\"],\n",
    "        \"DC1\": [\"XCR1\", \"CLEC9A\"],\n",
    "        \"DC2\": [\"CLEC10A\", \"THBD\"],\n",
    "        \"DC3\": [\"CSF3R\", \"CSF2RA\"],\n",
    "        \"mRegDC\": [\"LAMP3\", \"CCR7\"],\n",
    "        \"pDC\": [\"IL3RA\", \"KIT\", \"IRF8\"],\n",
    "        \"Mac\": [\n",
    "            \"C1QA\",\n",
    "            \"TREM2\",\n",
    "            \"SIGLEC1\",\n",
    "            \"CD68\",\n",
    "            \"LYVE1\",\n",
    "            \"ARG1\",\n",
    "            \"CD163\",\n",
    "        ],\n",
    "        \"prolif\": [\"UBE2C\",\"NUSAP1\",\"MKI67\"],\n",
    "        \"IFN\": [\"ISG15\", \"IFI27\", \"STAT1\"],\n",
    "        \"contamination\": [\"CD3D\"],\n",
    "    },\n",
    "    standard_scale=\"var\",\n",
    "    color_map=\"Blues\",\n",
    "    groupby=\"leiden_res1\",\n",
    ") "
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "57904e6f-fc71-4d57-9751-6454b27fd709",
   "metadata": {},
   "source": [
    "# Not as useful compared to the umap gene expression graphs\n",
    "\n",
    "sc.pl.dotplot(\n",
    "    rna_adata,\n",
    "    {\n",
    "        \"Immune cell\": [\n",
    "            \"CD3D\",\n",
    "            \"CD3E\",\n",
    "            \"CD4\",\n",
    "            \"CD8A\",\n",
    "            \"CCR7\",\n",
    "            \"SELL\",\n",
    "            \"CD27\",\n",
    "            \"ITGAM\",\n",
    "            \"S100A8\",\n",
    "        ],\n",
    "        \"prolif\": [\"MKI67\"],\n",
    "        \"IFN\": [\"ISG15\", \"IFI27\", \"STAT1\"],\n",
    "        \"KC Basal\": [\"KRT15\", \"C1orf56\"],\n",
    "        \"KC Differentiating\": [\"KRT10\"],\n",
    "        \"KC Cornified\": [\"LOR\", \"IVL\"],\n",
    "        \"KC Granular\": [\"KRT2\"],\n",
    "        \"KC stem/hair/stress\": [\"KRT16\", \"KRT17\", \"KRT6A\"],\n",
    "        \"KC nail/nail\": [\n",
    "            \"KRT6B\",\n",
    "            \"KRT79\",\n",
    "        ],\n",
    "        \"KC beta-catenin+\": [ \"C1orf56\"]\n",
    "    },\n",
    "    standard_scale=\"var\",\n",
    "    color_map=\"Blues\",\n",
    "    groupby=\"leiden_res1\",\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e6c10919-6adb-4230-acec-f0087b73ecd0",
   "metadata": {},
   "source": [
    "sc.pl.dotplot(\n",
    "    rna_adata,\n",
    "    {\n",
    "        \"T cell\": [\n",
    "            \"CD3D\",\n",
    "            \"CD3E\",\n",
    "            \"CD4\",\n",
    "            \"CD8A\",\n",
    "            \"CCR7\",\n",
    "            \"SELL\",\n",
    "            \"CD27\",\n",
    "        ],\n",
    "        \"Trm\": [\"CD69\", \"CRTAM\", \"S1PR1\"],\n",
    "        \"Treg\": [\"FOXP3\", \"PDCD1\", \"CTLA4\", \"IL2RA\"],\n",
    "        \"NK\": [ \"NCR1\", \"GZMB\",\"GNLY\", \"XCL2\"],\n",
    "        \"ILC\": [\"IL7R\"],\n",
    "        \"prolif\": [\"MKI67\"],\n",
    "        \"IFN\": [\"ISG15\", \"IFI27\", \"STAT1\"],\n",
    "        \"Contamination\": [\"ITGAM\", \"S100A8\"],\n",
    "        \"Endothelial\": [\"CLDN5\", \"PLVAP\", \"SPARCL1\"],\n",
    "        \"Fibroblast\": [\"PDGFRA\", \"PDGFRB\", \"DCN\"]\n",
    "    },\n",
    "    standard_scale=\"var\",\n",
    "    color_map=\"Blues\",\n",
    "    groupby=\"leiden_res1\",\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e69f74c3-ddcd-406f-996d-02b29fbd8fa9",
   "metadata": {},
   "source": [
    "\n",
    "new_cell_dict1 = {\n",
    "    \"0\": \"KC\", #Differentiating\n",
    "    \"1\": \"Melanocytes\", \n",
    "    \"2\": \"KC\",\n",
    "    \"3\": \"Fibroblast\", \n",
    "    \"4\": \"Immune\",  #T\n",
    "    \"5\": \"Endothelial cell\", \n",
    "    \"6\": \"Melanocytes\", \n",
    "    \"7\": \"Immune\",  #Mac\n",
    "    \"8\": \"Immune\", #T \n",
    "    \"9\": \"Fibroblast\", \n",
    "    \"10\": \"KC\", \n",
    "    \"11\": \"Immune\", #IFN T\n",
    "    \"12\": \"KC\", \n",
    "    \"13\": \"KC\", \n",
    "    \"14\": \"Immune\", #Treg\n",
    "    \"15\": \"Melanocytes\",\n",
    "    \"16\": \"KC\",\n",
    "    \"17\": \"KC\", \n",
    "    \"18\": \"KC\",\n",
    "    \"19\": \"Melanocytes\",\n",
    "    \"20\": \"KC\", \n",
    "    \"21\": \"KC\",\n",
    "    \"22\": \"Immune\", #DC\n",
    "    \"23\": \"\",\n",
    "    \"24\": \"Immune\", #NK\n",
    "    \"25\": \"Fibroblast\",\n",
    "    \"26\": \"Immune\", # T\n",
    "    \"27\": \"Endothelial cell\",\n",
    "    \"28\": \"Immune\", #NK\n",
    "    \"29\": \"\",\n",
    "    \"30\": \"Immune\", #IFN T\n",
    "    \"31\": \"\",\n",
    "    \"32\": \"\",\n",
    "}\n",
    "\n",
    "new_cell_dict1_2 = {\n",
    "    \"0\": \"KC Differentiating\",\n",
    "    \"1\": \"Melanocytes\", \n",
    "    \"2\": \"KC Basal\",\n",
    "    \"3\": \"Fibroblast\", \n",
    "    \"4\": \"T cell\", \n",
    "    \"5\": \"Endothelial cell\", \n",
    "    \"6\": \"Melanocytes\", \n",
    "    \"7\": \"Macrophage\", \n",
    "    \"8\": \"CD8+ T cell\", \n",
    "    \"9\": \"Fibroblast\", \n",
    "    \"10\": \"KC Granular\", \n",
    "    \"11\": \"\", \n",
    "    \"12\": \"KC Differentiating\", #? \n",
    "    \"13\": \"KC Cornified\", \n",
    "    \"14\": \"Treg\", #Treg\n",
    "    \"15\": \"Melanocytes\",\n",
    "    \"16\": \"KC stem/hair/stress\",\n",
    "    \"17\": \"KC nail\", \n",
    "    \"18\": \"\", #?\n",
    "    \"19\": \"Melanocytes\",\n",
    "    \"20\": \"KC Basal\", \n",
    "    \"21\": \"KC stem/hair/stress\",\n",
    "    \"22\": \"DC\", #DC\n",
    "    \"23\": \"\",\n",
    "    \"24\": \"NK\",\n",
    "    \"25\": \"Fibroblast\",\n",
    "    \"26\": \"\",\n",
    "    \"27\": \"Endothelial cell\",\n",
    "    \"28\": \"NK\",\n",
    "    \"29\": \"\",\n",
    "    \"30\": \"\",\n",
    "    \"31\": \"\",\n",
    "    \"32\": \"\",\n",
    "}\n",
    "\n",
    "rna_adata.obs[\"Level1_res1\"] = [new_cell_dict1[x] for x in rna_adata.obs[\"leiden_res1\"]]\n",
    "rna_adata.obs[\"Level2_res1\"] = [new_cell_dict1_2[x] for x in rna_adata.obs[\"leiden_res1\"]]\n",
    "\n",
    "#mdata = idata[idata.obs[\"Level1\"].isin([\"Imm_Myeloid\"])].copy()\n",
    "#tdata = idata[idata.obs[\"Level1\"].isin([\"Imm_T cell\"])].copy()\n",
    "#mdata = idata[idata.obs[\"Level1\"].isin([\"Imm_Myeloid\"])].copy()\n",
    "#tdata = idata[idata.obs[\"Level1\"].isin([\"Imm_T cell\"])].copy()\n",
    "\n",
    "sc.pl.umap(\n",
    "    rna_adata,\n",
    "    color=[\"Level1_res1\",\"Level2_res1\"],\n",
    ")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "7412845e-34ab-417c-bbe8-9ec3298c9d54",
   "metadata": {},
   "source": [
    "## subset KC"
   ]
  },
  {
   "cell_type": "code",
   "id": "58ac8f76-9b67-434e-97c8-fbe02f163a38",
   "metadata": {},
   "source": [
    "# Subset Immune Cells\n",
    "idata = rna_adata[rna_adata.obs.Level1_res1 == 'Immune']\n",
    "\n",
    "idata.var[\"mt\"] = idata.var_names.str.startswith(\"MT-\")\n",
    "sc.pp.calculate_qc_metrics(\n",
    "    idata, qc_vars=[\"mt\"], percent_top=None, log1p=False, inplace=True\n",
    ")\n",
    "sc.pp.normalize_total(idata, target_sum=1e4)\n",
    "sc.pp.log1p(idata)\n",
    "sc.pp.highly_variable_genes(idata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "37c0fd4c-a8cd-4b1c-9ef7-0e8b539d3fb7",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "92d5f837-38c1-4c5b-9f63-e2099e729da3",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "3cb3c99e-5eff-49d6-977a-9bbaa1b9f7ed",
   "metadata": {},
   "source": [
    "# 6, 14 are doublets with keratinocytes?? Express KRT14\n",
    "\n",
    "new_cell_dict = {\n",
    "    \"0\": \"\",\n",
    "    \"1\": \"\", #\n",
    "    \"2\": \"\",\n",
    "    \"3\": \"\", #T\n",
    "    \"4\": \"\", #Basal\n",
    "    \"5\": \"\", #DC\n",
    "    \"6\": \"\", #\n",
    "    \"7\": \"\", #T\n",
    "    \"8\": \"\", #\n",
    "    \"9\": \"\", #\n",
    "    \"10\": \"\", #\n",
    "    \"11\": \"Melanocytes\", #\n",
    "    \"12\": \"\", #\n",
    "    \"13\": \"\", #Basal\n",
    "    \"14\": \"\",\n",
    "    \"15\": \"\", #DC\n",
    "    \"16\": \"\",\n",
    "    \"17\": \"\", #\n",
    "    \"18\": \"\",\n",
    "    \"19\": \"\",\n",
    "    \"20\": \"\", #DC\n",
    "    \"21\": \"\", #stem/hair/stress\n",
    "    \"22\": \"\"\n",
    "}\n",
    "rna_adata.obs[\"Level1\"] = [new_cell_dict[x] for x in rna_adata.obs[\"leiden\"]]\n",
    "\n",
    "\n",
    "# 6, 14 are doublets with keratinocytes?? Express KRT14\n",
    "\n",
    "\n",
    "rna_adata.obs[\"Level1\"] = [new_cell_dict[x] for x in rna_adata.obs[\"leiden\"]]\n",
    "#rna_adata.obs[\"Level2\"] = [new_cell_dict2[x] for x in rna_adata.obs[\"leiden\"]]\n",
    "\n",
    "#mdata = idata[idata.obs[\"Level1\"].isin([\"Imm_Myeloid\"])].copy()\n",
    "#tdata = idata[idata.obs[\"Level1\"].isin([\"Imm_T cell\"])].copy()\n",
    "#mdata = idata[idata.obs[\"Level1\"].isin([\"Imm_Myeloid\"])].copy()\n",
    "#tdata = idata[idata.obs[\"Level1\"].isin([\"Imm_T cell\"])].copy()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "48badf2f-8da5-4b60-8d77-244ff0efc57f",
   "metadata": {},
   "source": [
    "rna_adata.obs[\"leiden_res1\"]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "3164d5e3-8045-43dd-95f2-08d387633760",
   "metadata": {},
   "source": [
    "rna_adata.write_h5ad(\"/QRISdata/Q1851/Prakrithi/Melanoma_scRNAseq/Mel_3samples_75pcs.h5ad\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bf43f4e9-dc17-4b75-bd44-b35a44126034",
   "metadata": {},
   "source": [
    "import anndata\n",
    "rna_adata=anndata.read_h5ad(\"/QRISdata/Q1851/Prakrithi/Melanoma_scRNAseq/Mel_3samples_75pcs.h5ad\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "1b291a89-9d42-4f61-a371-0fc08f7402ac",
   "metadata": {},
   "source": [
    "rna_adata.raw.X"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9465d413-0685-48fc-999e-ea41d31b5571",
   "metadata": {},
   "source": [],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
